{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load libraries and necessary data files, and define global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,\n",
      "    do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl,\n",
      "    intersect, is.unsorted, lapply, lengths, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unlist, unsplit\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: IRanges\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "Loading required package: XVector\n",
      "Loading required package: BSgenome\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: rtracklayer\n"
     ]
    }
   ],
   "source": [
    "#--- Load libraries and necessary data files, and define global variables\n",
    "library(data.tree)\n",
    "library(data.table)\n",
    "library(Biostrings)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(GenomicRanges)\n",
    "genome <- Hsapiens\n",
    "source(\"process_L1s.r\")\n",
    "source(\"mapsequence.r\")\n",
    "load(\"./data/chrmpd.rda\") # load chromosome probability distribution\n",
    "load(\"./data/L1RankTable.rda\")\n",
    "load(\"./data/geneann.rda\")\n",
    "trpd <- read.table(\"./data/L1truncpd.csv\",sep=\",\")\n",
    "tdpd <- read.table(\"./data/L1transdpd.csv\",sep=\",\")\n",
    "for (i in names(Hsapiens)[1:24]){ # load all chromosome map files\n",
    "        load(paste0(\"./data/root_maps/\",i,\".rda\"))\n",
    "}\n",
    "strdict<-c(\"+\",\"-\")\n",
    "names(strdict)<-c(1,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "update_chrom_map <- function(chrnm,chrMap,sites_chrm,sites_loci,l1s) {  \n",
    "    \n",
    "if (length(sites_chrm)==0){\n",
    "    return(chrMap)\n",
    "}\n",
    "    \n",
    "ict<-chrMap$ict\n",
    "icl<-chrMap$icl\n",
    "iot<-chrMap$iot\n",
    "iol<-chrMap$iol\n",
    "insites<-chrMap$insites \n",
    "    \n",
    "chrloci = sites_loci[sites_chrm==chrnm] # Get the sites where insertions occurred in the chromosome\n",
    "chrl1s = l1s[sites_chrm==chrnm] # Get the L1s elements which were inserted    \n",
    "    \n",
    "for (i in 1:length(chrloci)) { # Loop over the simulated insertion points\n",
    "    insites[which(is.na(insites))]<- -1 # Replace NA with -1\n",
    "    indx <- insites>chrloci[i] # Get indices of target sites which lie downstream of the point\n",
    "    insites[indx] <- insites[indx] + width(chrl1s[i]) # Shift the target sites by the length of the L1\n",
    "    l1_map <- mapSeq_SV(chrl1s[i]) # Map target sites in the L1\n",
    "    l1_map$insites <- l1_map$insites + chrloci[i] # Convert L1 loci to chromosome loci\n",
    "    insites <- rbind(insites,l1_map$insites) # Add target sites within L1 to chrom map\n",
    "    ict <- rbind(ict,l1_map$ict)\n",
    "    icl <- rbind(icl,l1_map$icl)\n",
    "    iot <- rbind(iot,l1_map$iot)\n",
    "    iol <- rbind(iol,l1_map$iol)\n",
    "}\n",
    "\n",
    "return(list(insites,ict,icl,iot,iol))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gen_sim <- function(genome,node,copyNum) {\n",
    "\n",
    "        sites_loci<-c() # Initialize arrays for storing simulated ins. site data\n",
    "        sites_chrm<-c()\n",
    "        sites_strand<-c()\n",
    "        sites_classes<-c()\n",
    "\n",
    "        #--- Sample chromosomes based on probability ranking. The data file chrmpd.rda \n",
    "        #--- must either be provided or generated by running 'get_sv_dist.r'.\n",
    "        #--- Here the ranking is provided by the number of 'TTTT' patterns.\n",
    "        chrmlist<-sample(x=names(genome)[1:24],copyNum,replace=TRUE,prob=chrmcnt[,1])\n",
    "        chrmlist<-table(chrmlist)\n",
    "\n",
    "        for (chrnm in names(chrmlist)) {\n",
    "\n",
    "\n",
    "                #cat(\"\\nChromosome: \",chrnm)\n",
    "            \n",
    "                map<-get(paste0(chrnm,\"Map\"))\n",
    "                map<-update_chrom_map(chrnm,map,node$sites_chrm,node$sites_loci,node$l1s)\n",
    "                ict<-map$ict\n",
    "                icl<-map$icl\n",
    "                iot<-map$iot\n",
    "                iol<-map$iol\n",
    "                insites<-map$insites\n",
    "            \n",
    "\n",
    "                chrcopyNum<-chrmlist[[chrnm]]\n",
    "\n",
    "                pd <- c(11.55*length(which(!is.na(ict))),\n",
    "                        7.25*length(which(!is.na(icl))),\n",
    "                        1.95*length(which(!is.na(iot))),\n",
    "                        1*length(which(!is.na(iol))))\n",
    "                pd <- (pd/sum(pd))*(1-ENifrc)\n",
    "                pd <- append(pd,ENifrc)\n",
    "                #cat(\"\\nSite class distribution:\\n\",pd)\n",
    "\n",
    "                #--- Generates insertion sites\n",
    "                classes <- sample(x = c(1:5),chrcopyNum,replace=TRUE,prob=pd)\n",
    "                sites <- rep(0,chrcopyNum)\n",
    "                strand <-rep(0,chrcopyNum)\n",
    "            \n",
    "                for (i in 1:chrcopyNum) {\n",
    "                        if (classes[i]==1) {\n",
    "                                tmp<-sample(c(1,2),1)\n",
    "                                sites[i] <- insites[ict[sample(c(1:length(which(!is.na(ict[,tmp])))),1),tmp]]\n",
    "                                strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==2) {\n",
    "                                tmp<-sample(c(1,2),1)\n",
    "                                sites[i] <- insites[icl[sample(c(1:length(which(!is.na(icl[,tmp])))),1),tmp]]\n",
    "                                strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==3) {\n",
    "                                tmp<-sample(c(1,2),1)\n",
    "                                sites[i] <- insites[iot[sample(c(1:length(which(!is.na(iot[,tmp])))),1),tmp]]\n",
    "                                strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==4) {\n",
    "                                tmp<-sample(c(1,2),1)\n",
    "                                sites[i] <- insites[iol[sample(c(1:length(which(!is.na(iol[,tmp])))),1),tmp]]\n",
    "                                strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==5) {\n",
    "                                sites[i]<-runif(1,1,length(genome[[chrnm]]))\n",
    "                                strand[i] <- strdict[[sample(c(1,2),1)]]\n",
    "                        }\n",
    "                }\n",
    "\n",
    "                sites_loci<-append(sites_loci,sites)\n",
    "                sites_chrm<-append(sites_chrm,rep(chrnm,chrcopyNum))\n",
    "                sites_strand<-append(sites_strand,strand)\n",
    "                sites_classes<-append(sites_classes,classes)\n",
    "\n",
    "        }\n",
    "\n",
    "        #--- Creates sequences for insertion\n",
    "        tmp <- process_L1s(genome,L1RankTable,trpd,tdpd,copyNum)\n",
    "        l1s <- tmp[[1]]\n",
    "        l1indcs <- tmp[[2]]\n",
    "        tdlen <- tmp[[3]]\n",
    "        trlen <- tmp[[4]]\n",
    "\n",
    "        return(list(l1s,sites_chrm,sites_loci,sites_strand))\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "rank_clone <- function(r, geneann, sites_chrm, sites_loci) {\n",
    "    \n",
    "    gene_hits=0; # set counter to zero\n",
    "    tsg_hits=0;\n",
    "    for (i in 1:length(sites_chrm)) { # loop over chromosomes inserted into\n",
    "        tmp=geneann[geneann$chrom==unique(sites_chrm)[i]] # reduce annotation table to entries for current chrom\n",
    "        chrmann_ntsg=tmp[tmp$istsg==0]\n",
    "        chrmann_tsg =tmp[tmp$istsg==1]\n",
    "        tmp = sites_loci[sites_chrm==unique(sites_chrm)[i]] # reduce insertion loci to entries for current chrom\n",
    "        tmp_hits = between(tmp,chrmann_ntsg$start,chrmann_ntsg$end) # create logical for insertions, whether into non-tsg-gene or not\n",
    "        gene_hits=gene_hits+length(which(tmp_hits==TRUE)) # count the number of non-tsg-gene insertions\n",
    "        tmp_hits  = between(tmp,chrmann_tsg$start,chrmann_tsg$end) \n",
    "        tsg_hits =tsg_hits+length(which(tmp_hits==TRUE))\n",
    "    }\n",
    "    \n",
    "    if (gene_hits > 0) {\n",
    "        r=0\n",
    "    } else if (tsg_hits > 0) {\n",
    "        r = r+tsg_hits*2; # TSG insertion doubles cell division rate\n",
    "    }\n",
    "    \n",
    "    return(r)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "maybeTranspose <- function(node,tnum) {\n",
    "    \n",
    "    if (sample(x=c(0,1),1,prob=c(1-node$tp, node$tp))) {\n",
    "        simout <- gen_sim(genome,node, 1)#round(runif(1,1,3)))\n",
    "        #ncln = ncln+1\n",
    "        r_tmp <- rank_clone(node$r, geneann, simout$sites_chrm, simout$sites_loci)\n",
    "        node$AddChild(tnum)\n",
    "        Set(node$children, r=r_tmp, tp=node$tp, l1s=node$l1s, sites=node$sites, ncells=1)\n",
    "        \n",
    "    }\n",
    "    \n",
    "    node$ncells <- node$ncells*(2^node$r)\n",
    " \n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--- Set simulation parameters\n",
    "ENifrc<- .1       # Fraction of endonuclease-independent (random) insertions\n",
    "inPopSize <- 10   # Initial number of cells in root clone\n",
    "inDivRate <- 1    # Initial division rate\n",
    "intp <- 0.5       # Initial probability of transposition\n",
    "\n",
    "NT <- 20           # Number of time iterations\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clone tree creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   user  system elapsed \n",
       "696.004 276.192 972.214 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "CellPop <- Node$new(1)\n",
    "CellPop$ncells <- inPopSize\n",
    "CellPop$r <- inDivRate\n",
    "CellPop$tp <- intp\n",
    "CellPop$l1s <- DNAStringSet()\n",
    "CellPop$sites_chrm <- c()\n",
    "CellPop$sites_loci <- c()\n",
    "ptm <- proc.time()\n",
    "for (i in 2:NT) {\n",
    "    \n",
    "    CellPop$Do(maybeTranspose,i)\n",
    "               \n",
    "}\n",
    "proc.time() - ptm\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                           levelName  ncells  r\n",
      "1   1                                                5242880  1\n",
      "2    ¦--2                                                 16  1\n",
      "3    ¦   ¦--4                                              2  1\n",
      "4    ¦   ¦   ¦--6                                          2  1\n",
      "5    ¦   ¦   ¦   ¦--7                                      8  1\n",
      "6    ¦   ¦   ¦   ¦   ¦--8                                  2  1\n",
      "7    ¦   ¦   ¦   ¦   ¦   ¦--12                             2  1\n",
      "8    ¦   ¦   ¦   ¦   ¦   ¦   ¦--15                         2  1\n",
      "9    ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--17                     2  1\n",
      "10   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                 4  1\n",
      "11   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20             1  1\n",
      "12   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--19                 2  1\n",
      "13   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                     2  1\n",
      "14   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "15   ¦   ¦   ¦   ¦   ¦   ¦   ¦--17                         2  1\n",
      "16   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "17   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                         2  1\n",
      "18   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "19   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                         2  1\n",
      "20   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "21   ¦   ¦   ¦   ¦   ¦   ¦   °--20                         1  1\n",
      "22   ¦   ¦   ¦   ¦   ¦   ¦--14                             2  1\n",
      "23   ¦   ¦   ¦   ¦   ¦   ¦   ¦--15                         4  1\n",
      "24   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                     2  1\n",
      "25   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "26   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                     2  1\n",
      "27   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "28   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "29   ¦   ¦   ¦   ¦   ¦   ¦   ¦--16                         4  1\n",
      "30   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                     2  1\n",
      "31   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "32   ¦   ¦   ¦   ¦   ¦   ¦   ¦--17                         4  1\n",
      "33   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                         4  1\n",
      "34   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                     2  1\n",
      "35   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "36   ¦   ¦   ¦   ¦   ¦   ¦   °--19                         2  1\n",
      "37   ¦   ¦   ¦   ¦   ¦   ¦--17                             2  1\n",
      "38   ¦   ¦   ¦   ¦   ¦   ¦   °--18                         4  1\n",
      "39   ¦   ¦   ¦   ¦   ¦   ¦--18                             2  1\n",
      "40   ¦   ¦   ¦   ¦   ¦   ¦   °--20                         1  1\n",
      "41   ¦   ¦   ¦   ¦   ¦   ¦--19                             2  1\n",
      "42   ¦   ¦   ¦   ¦   ¦   ¦   °--20                         1  1\n",
      "43   ¦   ¦   ¦   ¦   ¦   °--20                             1  1\n",
      "44   ¦   ¦   ¦   ¦   ¦--9                                  2  1\n",
      "45   ¦   ¦   ¦   ¦   ¦   ¦--12                             4  1\n",
      "46   ¦   ¦   ¦   ¦   ¦   ¦   ¦--13                         2  1\n",
      "47   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--17                     2  1\n",
      "48   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "49   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                     2  1\n",
      "50   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "51   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "52   ¦   ¦   ¦   ¦   ¦   ¦   ¦--14                         2  1\n",
      "53   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--16                     2  1\n",
      "54   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--17                 2  1\n",
      "55   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18             4  1\n",
      "56   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20         1  1\n",
      "57   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--19             2  1\n",
      "58   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦       °--20         1  1\n",
      "59   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                 2  1\n",
      "60   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19             2  1\n",
      "61   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20             1  1\n",
      "62   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                 2  1\n",
      "63   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "64   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--17                     2  1\n",
      "65   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--18                 4  1\n",
      "66   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦       ¦--19             2  1\n",
      "67   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦       ¦   °--20         1  1\n",
      "68   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦       °--20             1  1\n",
      "69   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                     2  1\n",
      "70   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--19                 2  1\n",
      "71   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                     2  1\n",
      "72   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "73   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "74   ¦   ¦   ¦   ¦   ¦   ¦   ¦--15                         2  1\n",
      "75   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--16                     4  1\n",
      "76   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                 2  1\n",
      "77   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20             1  1\n",
      "78   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                 2  1\n",
      "79   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20             1  1\n",
      "80   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "81   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--17                     4  1\n",
      "82   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "83   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--19                     2  1\n",
      "84   ¦   ¦   ¦   ¦   ¦   ¦   ¦--16                         2  1\n",
      "85   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                     2  1\n",
      "86   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--19                 2  1\n",
      "87   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "88   ¦   ¦   ¦   ¦   ¦   ¦   °--20                         1  1\n",
      "89   ¦   ¦   ¦   ¦   ¦   ¦--14                             4  1\n",
      "90   ¦   ¦   ¦   ¦   ¦   ¦   ¦--15                         2  1\n",
      "91   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--16                     2  1\n",
      "92   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                 1  1\n",
      "93   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--17                     2  1\n",
      "94   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "95   ¦   ¦   ¦   ¦   ¦   ¦   °--20                         1  1\n",
      "96   ¦   ¦   ¦   ¦   ¦   ¦--17                             4  1\n",
      "97   ¦   ¦   ¦   ¦   ¦   ¦   ¦--18                         2  1\n",
      "98   ¦   ¦   ¦   ¦   ¦   ¦   ¦   ¦--19                     2  1\n",
      "99   ¦   ¦   ¦   ¦   ¦   ¦   ¦   °--20                     1  1\n",
      "100  ¦   ¦   ¦   ¦   ¦   ¦   °--... 1 nodes w/ 0 sub      NA NA\n",
      "101  ¦   ¦   ¦   ¦   ¦   °--... 1 nodes w/ 2 sub          NA NA\n",
      "102  ¦   ¦   ¦   ¦   °--... 7 nodes w/ 113 sub            NA NA\n",
      "103  ¦   ¦   ¦   °--... 7 nodes w/ 361 sub                NA NA\n",
      "104  ¦   ¦   °--... 4 nodes w/ 397 sub                    NA NA\n",
      "105  ¦   °--... 11 nodes w/ 1624 sub                      NA NA\n",
      "106  °--... 4 nodes w/ 3184 sub                           NA NA\n"
     ]
    }
   ],
   "source": [
    "print(CellPop,'ncells','r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "save(CellPop,file=\"./data/CellPop_out.rda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
