{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load libraries and necessary data files, and define global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--- Load libraries and necessary data files, and define global variables\n",
    "library(data.tree)\n",
    "library(data.table)\n",
    "library(Biostrings)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(GenomicRanges)\n",
    "genome <- Hsapiens\n",
    "source(\"process_L1s.r\")\n",
    "source(\"mapsequence.r\")\n",
    "load(\"../data/chrmpd.rda\") # load chromosome probability distribution\n",
    "load(\"../data/L1RankTable.rda\")\n",
    "load(\"../data/exann.rda\")\n",
    "trpd <- read.table(\"../data/L1truncpd.csv\",sep=\",\")\n",
    "tdpd <- read.table(\"../data/L1transdpd.csv\",sep=\",\")\n",
    "for (i in names(Hsapiens)[1:24]){ # load all chromosome map files\n",
    "#         cat(paste0(\"Loading map file... \",i,\"\\n\"))\n",
    "        load(paste0(\"../data/root_maps/\",i,\".rda\"))\n",
    "}\n",
    "strdict<-c(\"+\",\"-\")\n",
    "names(strdict)<-c(1,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To simulate the transposition of an L1 sequence in a given genome\n",
    "#\n",
    "# INPUT:\n",
    "#   genome        (BSgenome) reference genome\n",
    "#   node          (data.tree node) input node of tree\n",
    "#   copyNum       (integer) number of L1 insertions to simulate\n",
    "#\n",
    "# OUTPUT: (list) inserted sequences, sites, and strand\n",
    "\n",
    "gen_sim <- function(genome,node,copyNum) {\n",
    "\n",
    "        sites_loci<-c() # Initialize arrays for storing simulated ins. site data\n",
    "        sites_chrm<-c()\n",
    "        sites_strand<-c()\n",
    "        sites_classes<-c()\n",
    "\n",
    "        #--- Sample chromosomes based on probability ranking. The data file chrmpd.rda \n",
    "        #--- must either be provided or generated by running 'get_sv_dist.r'.\n",
    "        #--- Here the ranking is provided by the number of 'TTTT' (closed-tight) sites.\n",
    "        chrmlist<-sample(x=names(genome)[1:24],copyNum,replace=TRUE,prob=chrmcnt[,1])\n",
    "        chrmlist<-table(chrmlist)\n",
    "\n",
    "        for (chrnm in names(chrmlist)) {\n",
    "\n",
    "\n",
    "#                 cat(\"\\nChromosome: \",chrnm)\n",
    "\n",
    "                map<-get(paste0(chrnm,\"Map\"))\n",
    "                map<-update_chrom_map(chrnm,map,node$tes[[2]],node$tes[[3]],node$tes[[1]])\n",
    "                ict<-map[[2]]\n",
    "                icl<-map[[3]]\n",
    "                iot<-map[[4]]\n",
    "                iol<-map[[5]]\n",
    "                insites<-map[[1]]\n",
    "\n",
    "\n",
    "                chrcopyNum<-chrmlist[[chrnm]]\n",
    "\n",
    "                pd <- c(11.55*length(which(!is.na(ict))),\n",
    "                        7.25*length(which(!is.na(icl))),\n",
    "                        1.95*length(which(!is.na(iot))),\n",
    "                        1*length(which(!is.na(iol))))\n",
    "                pd <- (pd/sum(pd))*(1-ENifrc)\n",
    "                pd <- append(pd,ENifrc)\n",
    "#                 cat(\"\\nSite class distribution:\\n\",pd)\n",
    "\n",
    "                #--- Generates insertion sites\n",
    "                classes <- sample(x = c(1:5),chrcopyNum,replace=TRUE,prob=pd)\n",
    "                sites <- rep(0,chrcopyNum)\n",
    "                strand <-rep(0,chrcopyNum)\n",
    "\n",
    "            \n",
    "                for (i in 1:chrcopyNum) {\n",
    "                        if (classes[i]==1) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[ict[sample(c(1:length(which(!is.na(ict[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==2) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[icl[sample(c(1:length(which(!is.na(icl[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==3) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[iot[sample(c(1:length(which(!is.na(iot[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==4) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[iol[sample(c(1:length(which(!is.na(iol[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==5) {\n",
    "                            sites[i]<-runif(1,1,length(genome[[chrnm]]))\n",
    "                            strand[i] <- strdict[[sample(c(1,2),1)]]\n",
    "                        }\n",
    "                }\n",
    "\n",
    "                sites_loci<-append(sites_loci,sites)\n",
    "                sites_chrm<-append(sites_chrm,rep(chrnm,chrcopyNum))\n",
    "                sites_strand<-append(sites_strand,strand)\n",
    "                sites_classes<-append(sites_classes,classes)\n",
    "\n",
    "        }\n",
    "\n",
    "        #--- Creates sequences for insertion\n",
    "        tmp <- process_L1s(genome,L1RankTable,trpd,tdpd,copyNum)\n",
    "        l1s <- tmp[[1]]\n",
    "        l1indcs <- tmp[[2]]\n",
    "        tdlen <- tmp[[3]]\n",
    "        trlen <- tmp[[4]]\n",
    "        return(list(l1s,sites_chrm,sites_loci,sites_strand,l1indcs))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To check whether the division rate of a new clone should be changed from its parent\n",
    "#\n",
    "# INPUT:\n",
    "#   r             (float) division rate\n",
    "#   anno       (data frame) Annotation of genes (i.e. chromosome   start   end)\n",
    "#   sites_chrm    (numeric vector) chromosomes containing L1 insertions for the clone\n",
    "#   sites_loci    (character vector) insertion positions (in the respective chromosome)\n",
    "#\n",
    "# OUTPUT: (float) possibly updated division rate\n",
    "\n",
    "rank_clone <- function(r, anno, sites_chrm, sites_loci, gainp, lossp) {\n",
    "\n",
    "    gene_hits=0; # set counter to zero\n",
    "    tsg_hits=0;\n",
    "    for (i in 1:length(unique(sites_chrm))) { # loop over chromosomes inserted into\n",
    "        tmp=anno[anno$chrom==unique(sites_chrm)[i],] # reduce annotation table to entries for current chrom\n",
    "        chrmann_ntsg=tmp[tmp$istsg==0,]\n",
    "        chrmann_tsg =tmp[tmp$istsg==1,]\n",
    "        tmp = sites_loci[sites_chrm==unique(sites_chrm)[i]] # reduce insertion loci to entries for current chrom\n",
    "        tmp_hits = between(tmp,chrmann_ntsg$start,chrmann_ntsg$end) # create logical for insertions, whether into non-tsg-gene or not\n",
    "        gene_hits=gene_hits+length(which(tmp_hits==TRUE)) # count the number of non-tsg-gene insertions\n",
    "#         print(gene_hits)\n",
    "        tmp_hits  = between(tmp,chrmann_tsg$start,chrmann_tsg$end) # same for tsg-gene insertions\n",
    "        tsg_hits =tsg_hits+length(which(tmp_hits==TRUE))\n",
    "    }\n",
    "\n",
    "    if (gene_hits > 0 || tsg_hits > 0){\n",
    "        r=r*(lossp^gene_hits)*(gainp^tsg_hits)\n",
    "    }\n",
    "\n",
    "#     if (r < 0.25) { # If the division rate is below 0.25, the clone stops growing\n",
    "#         r<-0\n",
    "#     }\n",
    "\n",
    "    return(r)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: Updates the gene annotation of the clone\n",
    "#\n",
    "# INPUT:\n",
    "#   anno         (data frame) Annotation of genes (i.e. chromosome   start   end)\n",
    "#   simout          (list of lists) gen_sim output\n",
    "#   tes             (list of lists) Node tes\n",
    "#\n",
    "# OUTPUT: anno\n",
    "\n",
    "update_anno <- function(anno, simout, tes) {\n",
    "    \n",
    "    tmp = mapply(append, simout, tes, SIMPLIFY = FALSE)\n",
    "    for (i in 1:length(tmp[[3]])) {\n",
    "        # Shift the start loci of genes with start loci beyond the insertion by the width of the L1\n",
    "        anno[anno$chrom==tmp[[2]][i] & anno$start>tmp[[3]][i],]$start <- anno[anno$chrom==tmp[[2]][i] & anno$start>tmp[[3]][i],]$start + width(tmp[[1]][i])  \n",
    "        # Shift the end loci of genes with start loci beyond the insertion by the width of the L1\n",
    "        anno[anno$chrom==tmp[[2]][i] & anno$start>tmp[[3]][i],]$end <- anno[anno$chrom==tmp[[2]][i] & anno$start>tmp[[3]][i],]$end + width(tmp[[1]][i])  \n",
    "        # Shift the end locus of any gene with only end locus beyond the insertion by the width of the L1\n",
    "        anno[anno$chrom==tmp[[2]][i] & anno$end>tmp[[3]][i] & anno$start<tmp[[3]][i],]$end <- anno[anno$chrom==tmp[[2]][i] & anno$end>tmp[[3]][i] & anno$start<tmp[[3]][i],]$end + width(tmp[[1]][i])        \n",
    "    }\n",
    "    return(anno) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To update the insertion site annotation of a chromosome for a clone.\n",
    "# The L1 insertions which have occurred in the clone will be accounted for\n",
    "#\n",
    "# INPUT:\n",
    "#   chrnm         (string) chromosome name\n",
    "#   chrMap        (list) chromosome annotation\n",
    "#   sites_chrm    (numeric vector) chromosomes containing L1 insertions for the clone\n",
    "#   sites_loci    (character vector) insertion positions (in the respective chromosome)\n",
    "#   l1s           (DNAStringSet) l1 sequences and orientation \n",
    "#\n",
    "# OUTPUT: (list) updated chromosome annotation\n",
    "\n",
    "update_chrom_map <- function(chrnm,chrMap,sites_chrm,sites_loci,l1s) {\n",
    "\n",
    "        if (length(which(sites_chrm==chrnm))==0){\n",
    "                return(chrMap)\n",
    "        }\n",
    "\n",
    "        ict<-chrMap[[2]]\n",
    "        icl<-chrMap[[3]]\n",
    "        iot<-chrMap[[4]]\n",
    "        iol<-chrMap[[5]]\n",
    "        insites<-chrMap[[1]]\n",
    "\n",
    "        chrloci = sites_loci[sites_chrm==chrnm] # Get the sites where insertions occurred in the chromosome\n",
    "        chrl1s = l1s[sites_chrm==chrnm] # Get the L1s elements which were inserted    \n",
    "    \n",
    "        for (i in 1:length(chrloci)) { # Loop over the simulated insertion sites\n",
    "                insites[which(is.na(insites))]<- -1 # Replace NA with -1\n",
    "                indx <- insites>chrloci[i] # Get indices of target sites which lie downstream of the site\n",
    "                insites[indx] <- insites[indx] + width(chrl1s[i]) # Shift the target sites by the length of the L1\n",
    "                l1_map <- mapsequence(chrl1s[i]) # Map target sites in the L1\n",
    "                l1_map$insites <- l1_map$insites + chrloci[i] # Convert L1 loci to chromosome loci\n",
    "                insites <- rbind(insites,l1_map$insites) # Add target sites within L1 to chrom map\n",
    "                ict <- rbind(ict,l1_map$ict)\n",
    "                icl <- rbind(icl,l1_map$icl)\n",
    "                iot <- rbind(iot,l1_map$iot)\n",
    "                iol <- rbind(iol,l1_map$iol)\n",
    "        }\n",
    "\n",
    "        return(list(insites,ict,icl,iot,iol))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To call gen_sim.r with some probability (probability of transposition, tp) for a clone at a time step\n",
    "#\n",
    "# INPUT:\n",
    "#   node          (data.tree node) current node of the data tree\n",
    "#   tnum          (integer) time step number\n",
    "#\n",
    "# OUTPUT: void\n",
    "maybeTranspose <- function(node, sd, sp) {\n",
    "    if (node$r==0){ # if the division rate of the clone is zero, skip the node\n",
    "        return()\n",
    "    }\n",
    "    \n",
    "    # increase the number of cells by the existing number * the division rate factor\n",
    "    nc <- node$ncells[length(node$ncells)] + round(node$ncells[length(node$ncells)]*node$r)\n",
    "    \n",
    "    # sample from binomial distribution for number of transpositions\n",
    "    if (nc < 4.2e9) {ntrans <- rbinom(1,nc,node$cellP)} # rbinom() fails for large n\n",
    "    else {ntrans <- nc*node$cellP} # If n is too large, use the expected number of events (mean of distribution)\n",
    "    if (ntrans > 0) {\n",
    "        simout <- gen_sim(genome,node,ntrans)\n",
    "        nc <- nc-ntrans\n",
    "        for (i in 1:ntrans) {\n",
    "            l<<-l+1\n",
    "            if (width(simout[[1]][i])>=6000) {clp_tmp <- min(node$cellP+rootCellP*L1RankTable$score[simout[[5]][i]], 1)} # Transposition P increases with intact L1 insertions     \n",
    "            else {clp_tmp <- node$cellP}\n",
    "            tmp <- update_anno(exann,lapply(simout,'[',i),node$tes)\n",
    "            r_tmp <- rank_clone(node$r, tmp, lapply(simout,'[',i)[[2]], lapply(simout,'[',i)[[3]], sd, sp)\n",
    "            tmp<-mapply(append, lapply(simout,'[',i), node$tes, SIMPLIFY = FALSE)\n",
    "            node$AddChild(l, cellP=clp_tmp, ncells=1, r=r_tmp, tes=tmp)\n",
    "        }\n",
    "    }   \n",
    "    node$ncells <- append(node$ncells,nc)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--- Set simulation parameters\n",
    "ENifrc<- .1       # Fraction of endonuclease-independent (random) insertions\n",
    "rootNCells <- 1   # Initial number of cells in root clone\n",
    "rootDivRate <- 1  # Initial division rate\n",
    "rootCellP <- 0.2      # Probability of transposition / timestep of a single cell\n",
    "\n",
    "NT <- 5           # Number of time steps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clone tree creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "l<-1 # Clone counter\n",
    "CellPop <- Node$new(1)\n",
    "CellPop$ncells <- c(rootNCells)\n",
    "CellPop$r <- rootDivRate\n",
    "CellPop$tes <- list(DNAStringSet(c(\"TTATTTA\")),c(\"chr1\"),c(1001140),c(\"+\"),c(0))\n",
    "CellPop$cellP <- rootCellP\n",
    "# CellPop$r <- rank_clone(CellPop$r, exann, CellPop$tes[[2]], CellPop$tes[[3]])\n",
    "# CellPop$r\n",
    "ptm <- proc.time()\n",
    "for (i in 2:NT) {\n",
    "    \n",
    "    CellPop$Do(maybeTranspose, 1.25, 0.8)\n",
    "    \n",
    "}\n",
    "proc.time() - ptm\n",
    "\n",
    "# print(CellPop,'ncells','r')\n",
    "# sum(vapply(CellPop$Get('ncells'),tail,n=1L,FUN.VALUE = numeric(1)))\n",
    "CellPop$Get('ncells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(CellPop,'r')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
