{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load libraries and necessary data files, and define global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--- Load libraries and necessary data files, and define global variables\n",
    "library(data.tree)\n",
    "library(data.table)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "genome <- Hsapiens\n",
    "load(\"../data/exann.rda\")\n",
    "load(\"../data/chrmpd.rda\") # load chromosome probability distribution\n",
    "load(\"../data/L1RankTable.rda\")\n",
    "load('../data/chromSitePd.rda')\n",
    "trpd <- read.table(\"../data/L1truncpd.csv\",sep=\",\")\n",
    "tdpd <- read.table(\"../data/L1transdpd.csv\",sep=\",\")\n",
    "for (i in names(Hsapiens)[1:24]){ # load all chromosome map files\n",
    "        load(paste0(\"../data/root_maps/\",i,\".rda\"))\n",
    "}\n",
    "strdict<-c(\"+\",\"-\")\n",
    "names(strdict)<-c(1,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To simulate L1 insertions in hg38\n",
    "#\n",
    "# INPUT:\n",
    "#   copyNum       (integer) number of L1 insertions to simulate\n",
    "#\n",
    "# OUTPUT:         (list) loci of insertions, indices of inserted L1s from L1RankTable\n",
    "# \n",
    "# Note: See gen-sim-notebook.ipynb for a more highly commented version\n",
    "gen_sim <- function(copyNum) {\n",
    "\n",
    "        sites_loci<-c() # Initialize arrays for storing simulated ins. site data\n",
    "        sites_chrm<-c()\n",
    "        sites_strand<-c()\n",
    "        sites_classes<-c()\n",
    "\n",
    "        #--- Sample chromosomes based on probability ranking. The data file chrmpd.rda \n",
    "        #--- must either be provided or generated by running 'get_sv_dist.r'.\n",
    "        #--- Here the ranking is provided by the number of 'TTTT' (closed-tight) sites.\n",
    "        chrmlist<-sample(x=names(genome)[1:24],copyNum,replace=TRUE,prob=chrmpd[,1])\n",
    "        chrmlist<-table(chrmlist)\n",
    "\n",
    "        for (chrnm in names(chrmlist)) {\n",
    "\n",
    "                map<-get(paste0(chrnm,\"Map\"))\n",
    "                ict<-map[[2]]\n",
    "                icl<-map[[3]]\n",
    "                iot<-map[[4]]\n",
    "                iol<-map[[5]]\n",
    "                insites<-map[[1]]\n",
    "\n",
    "                chrcopyNum<-chrmlist[[chrnm]]\n",
    "\n",
    "                #--- Generates insertion sites\n",
    "                classes <- sample(x = c(1:5),chrcopyNum,replace=TRUE,prob=pds[[chrnm]])\n",
    "                sites <- rep(0,chrcopyNum)\n",
    "                strand <-rep(0,chrcopyNum)\n",
    "\n",
    "            \n",
    "                for (i in 1:chrcopyNum) {\n",
    "                        if (classes[i]==1) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[ict[sample(c(1:length(which(!is.na(ict[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==2) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[icl[sample(c(1:length(which(!is.na(icl[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==3) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[iot[sample(c(1:length(which(!is.na(iot[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==4) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[iol[sample(c(1:length(which(!is.na(iol[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==5) {\n",
    "                            sites[i]<-runif(1,1,length(genome[[chrnm]]))\n",
    "                            strand[i] <- strdict[[sample(c(1,2),1)]]\n",
    "                        }\n",
    "                }\n",
    "\n",
    "                sites_loci<-append(sites_loci,sites)\n",
    "                sites_chrm<-append(sites_chrm,rep(chrnm,chrcopyNum))\n",
    "                sites_strand<-append(sites_strand,strand)\n",
    "                sites_classes<-append(sites_classes,classes)\n",
    "\n",
    "        }\n",
    "\n",
    "        # Sample copyNum L1s from the list (with replacement), based on activity ranking\n",
    "        l1indcs <- sample(x=c(1:40), copyNum, replace=TRUE, prob=L1RankTable$score[1:40])\n",
    "\n",
    "        # Sample copyNum truncation fractions and transduction lengths from their respective \n",
    "        # probability densities trpd, and tdpd\n",
    "#         trfrc <- sample(x=trpd[[1]], copyNum, replace=TRUE, prob=trpd[[2]])\n",
    "#         tdlen <- sample(x=tdpd[[1]], copyNum, replace=TRUE, prob=tdpd[[2]])\n",
    "\n",
    "    return(list(sites_chrm,sites_loci,sites_strand,l1indcs))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To count the number of driver and passenger mutations in a new clone and determine \n",
    "#          the birth rate and the genes affected\n",
    "#\n",
    "# INPUT:\n",
    "#   node          (data.tree node) current node of genotype tree\n",
    "#   anno          (data frame) Annotation of genes (i.e. chromosome   start   end)\n",
    "#   sites_chrm    (character vector) chromosomes containing L1 insertions for the clone\n",
    "#   sites_loci    (numeric vector) insertion positions (in the respective chromosome)\n",
    "#   sd            (float) driver mutation selection\n",
    "#   sp            (float) passenger mutation selection\n",
    "#\n",
    "# OUTPUT:\n",
    "#   B             (float) birth rate\n",
    "#   np            (int) number of passenger mutations\n",
    "#   nd            (int) number of driver mutations\n",
    "#   genes         (list of strings) list of symbols of affected genes\n",
    "count_hits <- function(node, anno, sites_chrm, sites_loci, sd, sp) {\n",
    "\n",
    "    np=0; # set counter to zero\n",
    "    nd=0;\n",
    "    genes=c();\n",
    "    \n",
    "    for (i in 1:length(unique(sites_chrm))) { # loop over chromosomes inserted into\n",
    "        \n",
    "        tmp=anno[anno$chrom==unique(sites_chrm)[i],] # reduce annotation table to entries for current chrom\n",
    "\n",
    "        tmp2 = sites_loci[sites_chrm==unique(sites_chrm)[i]] # reduce insertion loci to entries for current chrom\n",
    "        \n",
    "        ins <- lapply(tmp2,between,tmp$start,tmp$end) # find indices of affected genes\n",
    "        ins <- unlist(lapply(ins,which))\n",
    "        \n",
    "        if (length(ins)>0) {\n",
    "            genes <- append(genes,tmp$geneSym[ins])\n",
    "            np = np+length(which(tmp$istsg[ins]==0)) # count the number of non-tsg insertions\n",
    "            nd = nd+length(which(tmp$istsg[ins]==1)) # count the number of tsg insertions\n",
    "        } else {\n",
    "            np <- 0\n",
    "            nd <- 0\n",
    "        }\n",
    "\n",
    "    }\n",
    "    if (np > 0 || nd > 0) {\n",
    "        b <- ((1+sd)^(node$nd+nd))/((1+sp)^(node$np+np)) # Calculate birth rate of clone\n",
    "    } else {\n",
    "        b <- node$B\n",
    "    }\n",
    "\n",
    "    return(list(b, np, nd, genes))\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To call gen_sim.r with some probability (probability of transposition, mu) for a clone at a time step\n",
    "#\n",
    "# INPUT:\n",
    "#   node        (data.tree node) current node of the data tree\n",
    "#   sp          passenger mutation selection coefficient\n",
    "#   sd          driver mutation selection coefficient\n",
    "#\n",
    "# OUTPUT: void\n",
    "maybeTranspose <- function(node, sd, sp) {\n",
    "\n",
    "#     if (node$B==0){ # if the birth rate of the clone is zero, skip the node\n",
    "#         return()\n",
    "#     }\n",
    "    \n",
    "    # increase the number of cells by the existing number * the birth rate factor\n",
    "    inc <- round(tail(node$ncells,n=1)*node$B)\n",
    "    dec <- round(tail(node$ncells,n=1)*N/N0)\n",
    "    nc <- max(c(0,tail(node$ncells,n=1) + inc - dec))\n",
    "    N <<- N + (nc-tail(node$ncells,n=1)) # increment total population size\n",
    "    \n",
    "    # sample from binomial distribution for number of transpositions\n",
    "    if (nc < 4.2e9) {ntrans <- rbinom(1,nc,mu)} # rbinom() fails for large n\n",
    "    else {ntrans <- nc*mu} # If n is too large, use the expected number of events (mean of distribution)\n",
    "    if (ntrans > 0) {\n",
    "        simout <- gen_sim(ntrans)\n",
    "        nc <- nc-ntrans\n",
    "        for (i in 1:ntrans) { #### possible bottleneck\n",
    "            l<<-l+1\n",
    "            # determine birth rate, number of driver/passenger mutations, and affected genes of new clone\n",
    "            tmp1 <- count_hits(node, exann, lapply(simout,'[',i)[[1]], lapply(simout,'[',i)[[2]], sd, sp)\n",
    "            # determine the loci/L1s of insertions in the new clone\n",
    "            tmp2 <- mapply(append, lapply(simout,'[',i), node$tes, SIMPLIFY = FALSE)\n",
    "            node$AddChild(l, ncells=1, B=tmp1[[1]], nd=node$nd+tmp1[[3]], np=node$np+tmp1[[2]], genes=append(node$genes,tmp1[[4]]), tes=tmp2)\n",
    "            gn <<- append(gn,tmp1[[4]]) # append hit genes to list\n",
    "        }\n",
    "    }   \n",
    "    node$ncells <- append(node$ncells,nc)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--- Set simulation parameters\n",
    "ENifrc<- .1     # Fraction of endonuclease-independent (random) insertions\n",
    "N0 <- 100         # Initial number of cells in root clone\n",
    "B0 <- 1         # Initial birth rate\n",
    "mu <- 0.2         # Probability of transposition / timestep of a single cell\n",
    "\n",
    "NT <- 10         # Number of time steps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clone tree creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1"
      ],
      "text/latex": [
       "1"
      ],
      "text/markdown": [
       "1"
      ],
      "text/plain": [
       "[1] 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>elapsed:</strong> 10.1309999999999"
      ],
      "text/latex": [
       "\\textbf{elapsed:} 10.1309999999999"
      ],
      "text/markdown": [
       "**elapsed:** 10.1309999999999"
      ],
      "text/plain": [
       "elapsed \n",
       " 10.131 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       levelName                                  ncells\n",
      "1   1                            100, 74, 60, 45, 38, 31, 28, 24, 20, 17\n",
      "2    ¦--2                                      1, 1, 1, 1, 1, 1, 1, 1, 1\n",
      "3    ¦--3                                                           1, 0\n",
      "4    ¦   °--42                                                   1, 1, 0\n",
      "5    ¦       °--77                                      1, 1, 1, 1, 1, 1\n",
      "6    ¦--4                                                        1, 1, 0\n",
      "7    ¦   °--58                                                      1, 0\n",
      "8    ¦       °--78                                                  1, 0\n",
      "9    ¦           °--98                                     1, 1, 1, 1, 1\n",
      "10   ¦--5                                                        1, 1, 0\n",
      "11   ¦   °--59                                                      1, 0\n",
      "12   ¦       °--79                                               1, 1, 0\n",
      "13   ¦           °--115                                             1, 0\n",
      "14   ¦               °--131                                      1, 1, 1\n",
      "15   ¦--6                                      1, 1, 1, 1, 1, 1, 1, 1, 1\n",
      "16   ¦--7                                                  1, 1, 1, 1, 0\n",
      "17   ¦   °--99                                             1, 1, 1, 1, 0\n",
      "18   ¦       °--164                                                    1\n",
      "19   ¦--8                                                        1, 1, 0\n",
      "20   ¦   °--60                                             1, 1, 1, 1, 0\n",
      "21   ¦       °--132                                                 1, 0\n",
      "22   ¦           °--146                                             1, 1\n",
      "23   ¦--9                                                        1, 1, 0\n",
      "24   ¦   °--61                                                   1, 1, 0\n",
      "25   ¦       °--100                                        1, 1, 1, 1, 0\n",
      "26   ¦           °--165                                                1\n",
      "27   ¦--10                                                       1, 1, 0\n",
      "28   ¦   °--62                                                   1, 1, 0\n",
      "29   ¦       °--101                                        1, 1, 1, 1, 0\n",
      "30   ¦           °--166                                                1\n",
      "31   ¦--11                                                    1, 1, 1, 0\n",
      "32   ¦   °--80                                          1, 1, 1, 1, 1, 1\n",
      "33   ¦--12                                                    1, 1, 1, 0\n",
      "34   ¦   °--81                                          1, 1, 1, 1, 1, 1\n",
      "35   ¦--13                                     1, 1, 1, 1, 1, 1, 1, 1, 0\n",
      "36   ¦   °--167                                                        1\n",
      "37   ¦--14                                                 1, 1, 1, 1, 0\n",
      "38   ¦   °--102                                                  1, 1, 0\n",
      "39   ¦       °--133                                                 1, 0\n",
      "40   ¦           °--147                                             1, 1\n",
      "41   ¦--15                                                       1, 1, 0\n",
      "42   ¦   °--63                                                   1, 1, 0\n",
      "43   ¦       °--103                                              1, 1, 0\n",
      "44   ¦           °--134                                          1, 1, 1\n",
      "45   ¦--16                                     1, 1, 1, 1, 1, 1, 1, 1, 1\n",
      "46   ¦--17                                     1, 1, 1, 1, 1, 1, 1, 1, 1\n",
      "47   ¦--18                                                    1, 1, 1, 0\n",
      "48   ¦   °--82                                          1, 1, 1, 1, 1, 1\n",
      "49   ¦--19                                              1, 1, 1, 1, 1, 0\n",
      "50   ¦   °--116                                               1, 1, 1, 0\n",
      "51   ¦       °--168                                                    1\n",
      "52   ¦--20                                                 1, 1, 1, 1, 0\n",
      "53   ¦   °--104                                               1, 1, 1, 0\n",
      "54   ¦       °--148                                                 1, 1\n",
      "55   ¦--21                                     1, 1, 1, 1, 1, 1, 1, 1, 1\n",
      "56   ¦--22                                                       1, 1, 0\n",
      "57   ¦   °--64                                                1, 1, 1, 0\n",
      "58   ¦       °--117                                                 1, 0\n",
      "59   ¦           °--135                                          1, 1, 0\n",
      "60   ¦               °--169                                            1\n",
      "61   ¦--23                                                 1, 1, 1, 1, 0\n",
      "62   ¦   °--105                                                  1, 1, 0\n",
      "63   ¦       °--136                                              1, 1, 0\n",
      "64   ¦           °--170                                                1\n",
      "65   ¦--24                                                    1, 1, 1, 0\n",
      "66   ¦   °--83                                                   1, 1, 0\n",
      "67   ¦       °--118                                              1, 1, 0\n",
      "68   ¦           °--149                                             1, 1\n",
      "69   ¦--25                                                    1, 1, 1, 0\n",
      "70   ¦   °--84                                          1, 1, 1, 1, 1, 1\n",
      "71   ¦--26                                     1, 1, 1, 1, 1, 1, 1, 1, 0\n",
      "72   ¦   °--171                                                        1\n",
      "73   ¦--27                                                    1, 1, 1, 0\n",
      "74   ¦   °--85                                                      1, 0\n",
      "75   ¦       °--106                                        1, 1, 1, 1, 1\n",
      "76   ¦--28                                                          1, 0\n",
      "77   ¦   °--65                                       1, 1, 1, 1, 1, 1, 1\n",
      "78   ¦--29                                                          1, 0\n",
      "79   ¦   °--66                                                   1, 1, 0\n",
      "80   ¦       °--107                                                 1, 0\n",
      "81   ¦           °--119                                       1, 1, 1, 1\n",
      "82   ¦--30                                                       1, 1, 0\n",
      "83   ¦   °--86                                          1, 1, 1, 1, 1, 0\n",
      "84   ¦       °--172                                                    1\n",
      "85   ¦--31                                                          1, 0\n",
      "86   ¦   °--67                                       1, 1, 1, 1, 1, 1, 1\n",
      "87   ¦--32                                              1, 1, 1, 1, 1, 0\n",
      "88   ¦   °--137                                                     1, 0\n",
      "89   ¦       °--150                                                 1, 0\n",
      "90   ¦           °--173                                                1\n",
      "91   ¦--33                                                          1, 0\n",
      "92   ¦   °--68                                          1, 1, 1, 1, 1, 0\n",
      "93   ¦       °--151                                                 1, 1\n",
      "94   ¦--34                                        1, 1, 1, 1, 1, 1, 1, 1\n",
      "95   ¦--35                                                       1, 1, 0\n",
      "96   ¦   °--87                                          1, 1, 1, 1, 1, 1\n",
      "97   ¦--36                                        1, 1, 1, 1, 1, 1, 1, 0\n",
      "98   ¦   °--174                                                        1\n",
      "99   ¦--37                                              1, 1, 1, 1, 1, 0\n",
      "100  ¦   °--... 1 nodes w/ 0 sub                                        \n",
      "101  °--... 47 nodes w/ 35 sub                                          \n",
      "            B\n",
      "1   1.0000000\n",
      "2   1.0000000\n",
      "3   1.0000000\n",
      "4   1.0000000\n",
      "5   1.0000000\n",
      "6   1.0000000\n",
      "7   1.0000000\n",
      "8   1.0000000\n",
      "9   1.0000000\n",
      "10  1.0000000\n",
      "11  1.0000000\n",
      "12  1.0000000\n",
      "13  1.0000000\n",
      "14  1.0000000\n",
      "15  1.0000000\n",
      "16  1.0000000\n",
      "17  1.0000000\n",
      "18  1.0000000\n",
      "19  0.9090909\n",
      "20  0.9090909\n",
      "21  0.9090909\n",
      "22  0.9090909\n",
      "23  1.0000000\n",
      "24  1.0000000\n",
      "25  1.0000000\n",
      "26  1.0000000\n",
      "27  1.0000000\n",
      "28  1.0000000\n",
      "29  1.0000000\n",
      "30  1.0000000\n",
      "31  1.0000000\n",
      "32  1.0000000\n",
      "33  1.0000000\n",
      "34  1.0000000\n",
      "35  1.0000000\n",
      "36  1.0000000\n",
      "37  1.0000000\n",
      "38  1.0000000\n",
      "39  1.0000000\n",
      "40  1.0000000\n",
      "41  1.0000000\n",
      "42  1.0000000\n",
      "43  1.0000000\n",
      "44  0.9090909\n",
      "45  0.9090909\n",
      "46  0.9090909\n",
      "47  1.0000000\n",
      "48  1.0000000\n",
      "49  1.0000000\n",
      "50  1.0000000\n",
      "51  1.0000000\n",
      "52  1.0000000\n",
      "53  1.0000000\n",
      "54  1.0000000\n",
      "55  1.0000000\n",
      "56  1.0000000\n",
      "57  1.0000000\n",
      "58  1.0000000\n",
      "59  1.0000000\n",
      "60  1.0000000\n",
      "61  1.0000000\n",
      "62  1.0000000\n",
      "63  1.0000000\n",
      "64  1.0000000\n",
      "65  1.0000000\n",
      "66  1.0000000\n",
      "67  1.0000000\n",
      "68  1.0000000\n",
      "69  1.0000000\n",
      "70  1.0000000\n",
      "71  1.0000000\n",
      "72  1.0000000\n",
      "73  1.0000000\n",
      "74  1.0000000\n",
      "75  1.0000000\n",
      "76  1.0000000\n",
      "77  1.0000000\n",
      "78  1.0000000\n",
      "79  1.0000000\n",
      "80  1.0000000\n",
      "81  1.0000000\n",
      "82  1.0000000\n",
      "83  1.0000000\n",
      "84  1.0000000\n",
      "85  1.0000000\n",
      "86  1.0000000\n",
      "87  1.0000000\n",
      "88  1.0000000\n",
      "89  1.0000000\n",
      "90  1.0000000\n",
      "91  1.0000000\n",
      "92  1.0000000\n",
      "93  1.0000000\n",
      "94  1.0000000\n",
      "95  1.0000000\n",
      "96  1.0000000\n",
      "97  1.0000000\n",
      "98  1.0000000\n",
      "99  1.0000000\n",
      "100        NA\n",
      "101        NA\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "100"
      ],
      "text/latex": [
       "100"
      ],
      "text/markdown": [
       "100"
      ],
      "text/plain": [
       "[1] 100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li><strong>$`57511`</strong> = 'COG6'</li>\n",
       "\t<li><strong>$`7529`</strong> = 'YWHAB'</li>\n",
       "\t<li><strong>$`54471`</strong> = 'MIEF1'</li>\n",
       "\t<li><dl>\n",
       "\t<dt>$`23331`</dt>\n",
       "\t\t<dd>'TTC28'</dd>\n",
       "\t<dt>$`284900`</dt>\n",
       "\t\t<dd>'TTC28-AS1'</dd>\n",
       "</dl>\n",
       "</li>\n",
       "\t<li><strong>$`9837`</strong> = 'GINS1'</li>\n",
       "\t<li><dl>\n",
       "\t<dt>$`100847041`</dt>\n",
       "\t\t<dd>'MIR3680-2'</dd>\n",
       "\t<dt>$`388228`</dt>\n",
       "\t\t<dd>'SBK1'</dd>\n",
       "</dl>\n",
       "</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item \\textbf{\\$`57511`} = 'COG6'\n",
       "\\item \\textbf{\\$`7529`} = 'YWHAB'\n",
       "\\item \\textbf{\\$`54471`} = 'MIEF1'\n",
       "\\item \\begin{description}\n",
       "\\item[\\$`23331`] 'TTC28'\n",
       "\\item[\\$`284900`] 'TTC28-AS1'\n",
       "\\end{description}\n",
       "\n",
       "\\item \\textbf{\\$`9837`} = 'GINS1'\n",
       "\\item \\begin{description}\n",
       "\\item[\\$`100847041`] 'MIR3680-2'\n",
       "\\item[\\$`388228`] 'SBK1'\n",
       "\\end{description}\n",
       "\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. **$`57511`** = 'COG6'\n",
       "2. **$`7529`** = 'YWHAB'\n",
       "3. **$`54471`** = 'MIEF1'\n",
       "4. $`23331`\n",
       ":   'TTC28'\n",
       "$`284900`\n",
       ":   'TTC28-AS1'\n",
       "\n",
       "\n",
       "\n",
       "5. **$`9837`** = 'GINS1'\n",
       "6. $`100847041`\n",
       ":   'MIR3680-2'\n",
       "$`388228`\n",
       ":   'SBK1'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[[1]]$`57511`\n",
       "[1] \"COG6\"\n",
       "\n",
       "\n",
       "[[2]]\n",
       "[[2]]$`7529`\n",
       "[1] \"YWHAB\"\n",
       "\n",
       "\n",
       "[[3]]\n",
       "[[3]]$`54471`\n",
       "[1] \"MIEF1\"\n",
       "\n",
       "\n",
       "[[4]]\n",
       "[[4]]$`23331`\n",
       "[1] \"TTC28\"\n",
       "\n",
       "[[4]]$`284900`\n",
       "[1] \"TTC28-AS1\"\n",
       "\n",
       "\n",
       "[[5]]\n",
       "[[5]]$`9837`\n",
       "[1] \"GINS1\"\n",
       "\n",
       "\n",
       "[[6]]\n",
       "[[6]]$`100847041`\n",
       "[1] \"MIR3680-2\"\n",
       "\n",
       "[[6]]$`388228`\n",
       "[1] \"SBK1\"\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "l<-1 # Clone counter\n",
    "\n",
    "N <- N0 # total population size\n",
    "gn <- c() # list of genes hit\n",
    "\n",
    "CellPop <- Node$new(1) # Initialize data.tree as single node\n",
    "\n",
    "CellPop$ncells <- c(N0) # Set initial number of cells of clone\n",
    "CellPop$B <- B0 # Set initial birth rate of clone\n",
    "CellPop$np <- 0 # Set initial number of drivier mutations\n",
    "CellPop$nd <- 0 # Set initial number of passenger mutations\n",
    "CellPop$genes <- c()\n",
    "# Set the L1 insertions existing in root node\n",
    "CellPop$tes <- list(c(),c(),c(),c())\n",
    "# CellPop$tes <- list(c(\"chr1\"),c(69091),c(\"+\"),c(0))\n",
    "# CellPop$tes <- list(c(\"chr1\",\"chr1\"),c(11874,69091),c(\"+\",\"+\"),c(0,0))\n",
    "\n",
    "# update attributes based on initialized insertions\n",
    "tmp <- count_hits(CellPop, exann, CellPop$tes[[1]], CellPop$tes[[2]], 0.1, 0.1)\n",
    "CellPop$B <- tmp[[1]]\n",
    "CellPop$np <- c(tmp[[2]])\n",
    "CellPop$nd <- c(tmp[[3]])\n",
    "CellPop$genes <- c(tmp[[4]])\n",
    "\n",
    "# print birth rate of root clone\n",
    "CellPop$B\n",
    "\n",
    "ptm <- proc.time()\n",
    "for (i in 2:NT) {\n",
    "\n",
    "    t <- Traverse(CellPop,traversal='pre-order',filterFun=function(x) tail(x$ncells,n=1) > 0)\n",
    "    lapply(t,maybeTranspose,0.1,0.1)\n",
    "    \n",
    "}\n",
    "(proc.time() - ptm)[3]\n",
    "\n",
    "print(CellPop,'ncells','B')\n",
    "# sum(vapply(CellPop$Get('ncells'),tail,n=1L,FUN.VALUE = numeric(1)))\n",
    "# CellPop$Get('ncells')\n",
    "N\n",
    "gn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
