{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load libraries and necessary data files, and define global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,\n",
      "    do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl,\n",
      "    intersect, is.unsorted, lapply, lengths, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unlist, unsplit\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: IRanges\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "Loading required package: XVector\n",
      "Loading required package: BSgenome\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: rtracklayer\n"
     ]
    }
   ],
   "source": [
    "#--- Load libraries and necessary data files, and define global variables\n",
    "library(data.tree)\n",
    "library(data.table)\n",
    "library(Biostrings)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(GenomicRanges)\n",
    "genome <- Hsapiens\n",
    "load(\"../data/exann.rda\")\n",
    "load(\"../data/chrmpd.rda\") # load chromosome probability distribution\n",
    "load(\"../data/L1RankTable.rda\")\n",
    "load('../data/chromSitePd.rda')\n",
    "trpd <- read.table(\"../data/L1truncpd.csv\",sep=\",\")\n",
    "tdpd <- read.table(\"../data/L1transdpd.csv\",sep=\",\")\n",
    "for (i in names(Hsapiens)[1:24]){ # load all chromosome map files\n",
    "        load(paste0(\"../data/root_maps/\",i,\".rda\"))\n",
    "}\n",
    "strdict<-c(\"+\",\"-\")\n",
    "names(strdict)<-c(1,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To simulate the transposition of an L1 sequence in a given genome\n",
    "#\n",
    "# INPUT:\n",
    "#   copyNum       (integer) number of L1 insertions to simulate\n",
    "#\n",
    "# OUTPUT: (list) inserted sequences, sites, and strand\n",
    "\n",
    "gen_sim <- function(copyNum) {\n",
    "\n",
    "        sites_loci<-c() # Initialize arrays for storing simulated ins. site data\n",
    "        sites_chrm<-c()\n",
    "        sites_strand<-c()\n",
    "        sites_classes<-c()\n",
    "\n",
    "        #--- Sample chromosomes based on probability ranking. The data file chrmpd.rda \n",
    "        #--- must either be provided or generated by running 'get_sv_dist.r'.\n",
    "        #--- Here the ranking is provided by the number of 'TTTT' (closed-tight) sites.\n",
    "        chrmlist<-sample(x=names(genome)[1:24],copyNum,replace=TRUE,prob=chrmcnt[,1])\n",
    "        chrmlist<-table(chrmlist)\n",
    "\n",
    "        for (chrnm in names(chrmlist)) {\n",
    "\n",
    "                map<-get(paste0(chrnm,\"Map\"))\n",
    "                ict<-map[[2]]\n",
    "                icl<-map[[3]]\n",
    "                iot<-map[[4]]\n",
    "                iol<-map[[5]]\n",
    "                insites<-map[[1]]\n",
    "\n",
    "                chrcopyNum<-chrmlist[[chrnm]]\n",
    "\n",
    "                #--- Generates insertion sites\n",
    "                classes <- sample(x = c(1:5),chrcopyNum,replace=TRUE,prob=pds[[chrnm]])\n",
    "                sites <- rep(0,chrcopyNum)\n",
    "                strand <-rep(0,chrcopyNum)\n",
    "\n",
    "            \n",
    "                for (i in 1:chrcopyNum) {\n",
    "                        if (classes[i]==1) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[ict[sample(c(1:length(which(!is.na(ict[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==2) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[icl[sample(c(1:length(which(!is.na(icl[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==3) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[iot[sample(c(1:length(which(!is.na(iot[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==4) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[iol[sample(c(1:length(which(!is.na(iol[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==5) {\n",
    "                            sites[i]<-runif(1,1,length(genome[[chrnm]]))\n",
    "                            strand[i] <- strdict[[sample(c(1,2),1)]]\n",
    "                        }\n",
    "                }\n",
    "\n",
    "                sites_loci<-append(sites_loci,sites)\n",
    "                sites_chrm<-append(sites_chrm,rep(chrnm,chrcopyNum))\n",
    "                sites_strand<-append(sites_strand,strand)\n",
    "                sites_classes<-append(sites_classes,classes)\n",
    "\n",
    "        }\n",
    "\n",
    "        # Sample copyNum L1s from the list (with replacement), based on activity ranking\n",
    "        l1indcs <- sample(x=c(1:40), copyNum, replace=TRUE, prob=L1RankTable$score[1:40])\n",
    "\n",
    "        # Sample copyNum truncation fractions and transduction lengths from their respective \n",
    "        # probability densities trpd, and tdpd\n",
    "#         trfrc <- sample(x=trpd[[1]], copyNum, replace=TRUE, prob=trpd[[2]])\n",
    "#         tdlen <- sample(x=tdpd[[1]], copyNum, replace=TRUE, prob=tdpd[[2]])\n",
    "\n",
    "    return(list(sites_chrm,sites_loci,sites_strand,l1indcs))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To check whether the division rate of a new clone should be changed from its parent\n",
    "#\n",
    "# INPUT:\n",
    "#   r             (float) division rate\n",
    "#   anno          (data frame) Annotation of genes (i.e. chromosome   start   end)\n",
    "#   sites_chrm    (numeric vector) chromosomes containing L1 insertions for the clone\n",
    "#   sites_loci    (character vector) insertion positions (in the respective chromosome)\n",
    "#   gainp         (factor by which the division rate changes if a non-TSG is disrupted)\n",
    "#   lossp         (factor by which the division rate changes if a TSG is disrupted)\n",
    "#\n",
    "# OUTPUT: (float) possibly updated division rate\n",
    "rank_clone <- function(r, anno, sites_chrm, sites_loci, gainp, lossp) {\n",
    "\n",
    "    gene_hits=0; # set counter to zero\n",
    "    tsg_hits=0;\n",
    "    genes=c();\n",
    "    \n",
    "    for (i in 1:length(unique(sites_chrm))) { # loop over chromosomes inserted into\n",
    "        \n",
    "        tmp=anno[anno$chrom==unique(sites_chrm)[i],] # reduce annotation table to entries for current chrom\n",
    "\n",
    "        tmp2 = sites_loci[sites_chrm==unique(sites_chrm)[i]] # reduce insertion loci to entries for current chrom\n",
    "        \n",
    "        ins <- lapply(tmp2,between,tmp$start,tmp$end) # create logical for insertions, whether into non-tsg or not\n",
    "        ins <- unlist(lapply(ins,which))\n",
    "        \n",
    "        if (length(ins)>0) {\n",
    "            genes <- append(genes,tmp$geneSym[ins])\n",
    "        }\n",
    "        \n",
    "        gene_hits=gene_hits+length(which(tmp$istsg[ins]==0)) # count the number of non-tsg insertions\n",
    "        tsg_hits =tsg_hits +length(which(tmp$istsg[ins]==1)) # count the number of tsg insertions\n",
    "\n",
    "    }\n",
    "\n",
    "    if (gene_hits > 0 || tsg_hits > 0){\n",
    "        r=r*(lossp^gene_hits)*(gainp^tsg_hits)\n",
    "    }\n",
    "\n",
    "    return(list(r,genes))\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To call gen_sim.r with some probability (probability of transposition, tp) for a clone at a time step\n",
    "#\n",
    "# INPUT:\n",
    "#   node        (data.tree node) current node of the data tree\n",
    "#   sp          passenger mutation selection coefficient (<1)\n",
    "#   sd          driver mutation selection coefficient (>1)\n",
    "#\n",
    "# OUTPUT: void\n",
    "maybeTranspose <- function(node, sd, sp) {\n",
    "\n",
    "    if (node$r[[1]]==0){ # if the division rate of the clone is zero, skip the node\n",
    "        return()\n",
    "    }\n",
    "    \n",
    "    # increase the number of cells by the existing number * the division rate factor\n",
    "    nc <- node$ncells[length(node$ncells)] + round(node$ncells[length(node$ncells)]*node$r[[1]])\n",
    "    \n",
    "    # sample from binomial distribution for number of transpositions\n",
    "    if (nc < 4.2e9) {ntrans <- rbinom(1,nc,cellP)} # rbinom() fails for large n\n",
    "    else {ntrans <- nc*cellP} # If n is too large, use the expected number of events (mean of distribution)\n",
    "    if (ntrans > 0) {\n",
    "        simout <- gen_sim(ntrans)\n",
    "        nc <- nc-ntrans\n",
    "        for (i in 1:ntrans) {\n",
    "            l<<-l+1\n",
    "            r_tmp <- rank_clone(node$r, exann, lapply(simout,'[',i)[[1]], lapply(simout,'[',i)[[2]], sd, sp)\n",
    "            tmp<-mapply(append, lapply(simout,'[',i), node$tes, SIMPLIFY = FALSE)\n",
    "            node$AddChild(l, ncells=1, r=r_tmp[[1]], genes=append(node$genes,r_tmp[[2]]), tes=tmp)\n",
    "        }\n",
    "    }   \n",
    "    node$ncells <- append(node$ncells,nc)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--- Set simulation parameters\n",
    "ENifrc<- .1       # Fraction of endonuclease-independent (random) insertions\n",
    "rootNCells <- 1   # Initial number of cells in root clone\n",
    "rootDivRate <- 1  # Initial division rate\n",
    "cellP <- 1      # Probability of transposition / timestep of a single cell\n",
    "\n",
    "NT <- 3           # Number of time steps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clone tree creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.64"
      ],
      "text/latex": [
       "0.64"
      ],
      "text/markdown": [
       "0.64"
      ],
      "text/plain": [
       "[1] 0.64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>elapsed:</strong> 0.159999999999997"
      ],
      "text/latex": [
       "\\textbf{elapsed:} 0.159999999999997"
      ],
      "text/markdown": [
       "**elapsed:** 0.159999999999997"
      ],
      "text/plain": [
       "elapsed \n",
       "   0.16 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "3"
      ],
      "text/latex": [
       "3"
      ],
      "text/markdown": [
       "3"
      ],
      "text/plain": [
       "[1] 3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "l<-1 # Clone counter\n",
    "\n",
    "CellPop <- Node$new(1)\n",
    "CellPop$ncells <- c(rootNCells)\n",
    "CellPop$r <- rootDivRate\n",
    "CellPop$tes <- list(c(),c(),c(),c())\n",
    "# CellPop$tes <- list(c(\"chr1\"),c(69091),c(\"+\"),c(0))\n",
    "CellPop$tes <- list(c(\"chr1\",\"chr1\"),c(11874,69091),c(\"+\",\"+\"),c(0,0))\n",
    "\n",
    "rcout <- rank_clone(CellPop$r, exann, CellPop$tes[[1]], CellPop$tes[[2]],1.25,0.8)\n",
    "CellPop$r <- rcout[[1]]\n",
    "CellPop$genes <- c(rcout[[2]])\n",
    "CellPop$r\n",
    "\n",
    "ptm <- proc.time()\n",
    "for (i in 2:NT) {\n",
    "\n",
    "    CellPop$Do(maybeTranspose, 1.25, 0.8)\n",
    "    \n",
    "}\n",
    "(proc.time() - ptm)[3]\n",
    "\n",
    "# print(CellPop,'ncells','r','genes')\n",
    "sum(vapply(CellPop$Get('ncells'),tail,n=1L,FUN.VALUE = numeric(1)))\n",
    "# CellPop$Get('ncells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
