{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load libraries and necessary data files, and define global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BSgenome\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,\n",
      "    do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl,\n",
      "    intersect, is.unsorted, lapply, lengths, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unlist, unsplit\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: IRanges\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: Biostrings\n",
      "Loading required package: XVector\n",
      "Loading required package: rtracklayer\n",
      "'BiocParallel' did not register default BiocParallelParams:\n",
      "  invalid class “MulticoreParam” object: 1: ‘cluster’, ‘.clusterargs’, ‘RNGseed’ must be length 1\n",
      "invalid class “MulticoreParam” object: 2: ‘timeout’, ‘log’ must be length 1\n"
     ]
    }
   ],
   "source": [
    "#--- Load libraries and necessary data files, and define global variables\n",
    "library(data.tree)\n",
    "library(data.table)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "genome <- Hsapiens\n",
    "load(\"../data/exann.rda\")\n",
    "load(\"../data/chrmpd.rda\") # load chromosome probability distribution\n",
    "load(\"../data/L1RankTable.rda\")\n",
    "load('../data/chromSitePd.rda')\n",
    "trpd <- read.table(\"../data/L1truncpd.csv\",sep=\",\")\n",
    "tdpd <- read.table(\"../data/L1transdpd.csv\",sep=\",\")\n",
    "for (i in names(Hsapiens)[1:24]){ # load all chromosome map files\n",
    "        load(paste0(\"../data/root_maps/\",i,\".rda\"))\n",
    "}\n",
    "strdict<-c(\"+\",\"-\")\n",
    "names(strdict)<-c(1,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To simulate L1 insertions in hg38\n",
    "#\n",
    "# INPUT:\n",
    "#   copyNum       (integer) number of L1 insertions to simulate\n",
    "#\n",
    "# OUTPUT:         (list) loci of insertions, indices of inserted L1s from L1RankTable\n",
    "# \n",
    "# Note: See gen-sim-notebook.ipynb for a more highly commented version\n",
    "gen_sim <- function(copyNum) {\n",
    "\n",
    "        sites_loci<-c() # Initialize arrays for storing simulated ins. site data\n",
    "        sites_chrm<-c()\n",
    "        sites_strand<-c()\n",
    "        sites_classes<-c()\n",
    "\n",
    "        #--- Sample chromosomes based on probability ranking. The data file chrmpd.rda \n",
    "        #--- must either be provided or generated by running 'get_sv_dist.r'.\n",
    "        #--- Here the ranking is provided by the number of 'TTTT' (closed-tight) sites.\n",
    "        chrmlist<-sample(x=names(genome)[1:24],copyNum,replace=TRUE,prob=chrmpd[,1])\n",
    "        chrmlist<-table(chrmlist)\n",
    "\n",
    "        for (chrnm in names(chrmlist)) {\n",
    "\n",
    "                map<-get(paste0(chrnm,\"Map\"))\n",
    "                ict<-map[[2]]\n",
    "                icl<-map[[3]]\n",
    "                iot<-map[[4]]\n",
    "                iol<-map[[5]]\n",
    "                insites<-map[[1]]\n",
    "\n",
    "                chrcopyNum<-chrmlist[[chrnm]]\n",
    "\n",
    "                #--- Generates insertion sites\n",
    "                classes <- sample(x = c(1:5),chrcopyNum,replace=TRUE,prob=pds[[chrnm]])\n",
    "                sites <- rep(0,chrcopyNum)\n",
    "                strand <-rep(0,chrcopyNum)\n",
    "\n",
    "            \n",
    "                for (i in 1:chrcopyNum) {\n",
    "                        if (classes[i]==1) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[ict[sample(c(1:length(which(!is.na(ict[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==2) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[icl[sample(c(1:length(which(!is.na(icl[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==3) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[iot[sample(c(1:length(which(!is.na(iot[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==4) {\n",
    "                            tmp<-sample(c(1,2),1)\n",
    "                            sites[i] <- insites[iol[sample(c(1:length(which(!is.na(iol[,tmp])))),1),tmp]]\n",
    "                            strand[i] <- strdict[[tmp]]\n",
    "                        } else if (classes[i]==5) {\n",
    "                            sites[i]<-runif(1,1,length(genome[[chrnm]]))\n",
    "                            strand[i] <- strdict[[sample(c(1,2),1)]]\n",
    "                        }\n",
    "                }\n",
    "\n",
    "                sites_loci<-append(sites_loci,sites)\n",
    "                sites_chrm<-append(sites_chrm,rep(chrnm,chrcopyNum))\n",
    "                sites_strand<-append(sites_strand,strand)\n",
    "                sites_classes<-append(sites_classes,classes)\n",
    "\n",
    "        }\n",
    "\n",
    "        # Sample copyNum L1s from the list (with replacement), based on activity ranking\n",
    "        l1indcs <- sample(x=c(1:40), copyNum, replace=TRUE, prob=L1RankTable$score[1:40])\n",
    "\n",
    "        # Sample copyNum truncation fractions and transduction lengths from their respective \n",
    "        # probability densities trpd, and tdpd\n",
    "#         trfrc <- sample(x=trpd[[1]], copyNum, replace=TRUE, prob=trpd[[2]])\n",
    "#         tdlen <- sample(x=tdpd[[1]], copyNum, replace=TRUE, prob=tdpd[[2]])\n",
    "\n",
    "    return(list(sites_chrm,sites_loci,sites_strand,l1indcs))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To count the number of driver and passenger mutations in a new clone and determine \n",
    "#          the birth rate and the genes affected\n",
    "#\n",
    "# INPUT:\n",
    "#   node          (data.tree node) current node of genotype tree\n",
    "#   anno          (data frame) Annotation of genes (i.e. chromosome   start   end)\n",
    "#   sites_chrm    (character vector) chromosomes containing L1 insertions for the clone\n",
    "#   sites_loci    (numeric vector) insertion positions (in the respective chromosome)\n",
    "#   sd            (float) driver mutation selection\n",
    "#   sp            (float) passenger mutation selection\n",
    "#\n",
    "# OUTPUT:\n",
    "#   B             (float) birth rate\n",
    "#   np            (int) number of passenger mutations\n",
    "#   nd            (int) number of driver mutations\n",
    "#   genes         (list of strings) list of symbols of affected genes\n",
    "count_hits <- function(node, anno, sites_chrm, sites_loci, sd, sp) {\n",
    "\n",
    "    np<-0; # set counters to zero\n",
    "    nd<-0;\n",
    "    genes <- c()\n",
    "    \n",
    "    for (i in 1:length(unique(sites_chrm))) { # loop over chromosomes inserted into\n",
    "        \n",
    "        tmp=anno[anno$chrom==unique(sites_chrm)[i],] # reduce annotation table to entries for current chrom\n",
    "\n",
    "        tmp2 = sites_loci[sites_chrm==unique(sites_chrm)[i]] # reduce insertion loci to entries for current chrom\n",
    "        \n",
    "        ins <- lapply(tmp2,between,tmp$start,tmp$end) # find indices of affected genes\n",
    "        ins <- unlist(lapply(ins,which))\n",
    "        \n",
    "        if (length(ins)>0) {\n",
    "            genes <- rbind(genes,data.table(sym=tmp$geneSym[ins],tsg=tmp$istsg[ins]))\n",
    "            np <- np+length(which(tmp$istsg[ins]==0)) # count the number of non-tsg insertions\n",
    "            nd <- nd+length(which(tmp$istsg[ins]==1)) # count the number of tsg insertions\n",
    "        } else {\n",
    "            np <- 0\n",
    "            nd <- 0\n",
    "        }\n",
    "\n",
    "    }\n",
    "    if (np > 0 || nd > 0) {\n",
    "        b <- ((1+sd)^(node$nd+nd))/((1+sp)^(node$np+np)) # Calculate birth rate of clone\n",
    "    } else {\n",
    "        b <- node$B\n",
    "    }\n",
    "\n",
    "    return(list(b, np, nd, genes))\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PURPOSE: To call gen_sim.r with some probability (probability of transposition, mu) for a clone at a time step\n",
    "#\n",
    "# INPUT:\n",
    "#   node        (data.tree node) current node of the data tree\n",
    "#   sp          passenger mutation selection coefficient\n",
    "#   sd          driver mutation selection coefficient\n",
    "#\n",
    "# OUTPUT: void\n",
    "maybeTranspose <- function(node, sd, sp) {\n",
    "\n",
    "    if (node$B==0){ # if the birth rate of the clone is zero, skip the node\n",
    "        return()\n",
    "    }\n",
    "    \n",
    "    # increase the number of cells by the existing number * the birth rate factor\n",
    "    nc <- tail(node$ncells,n=1) + round(tail(node$ncells,n=1)*node$B)\n",
    "    \n",
    "    # sample from binomial distribution for number of transpositions\n",
    "    if (nc < 4.2e9) {ntrans <- rbinom(1,nc,mu)} # rbinom() fails for large n\n",
    "    else {ntrans <- nc*mu} # If n is too large, use the expected number of events (mean of distribution)\n",
    "    if (ntrans > 0) {\n",
    "        simout <- gen_sim(ntrans)\n",
    "        nc <- nc-ntrans\n",
    "        for (i in 1:ntrans) { #### possible bottleneck\n",
    "            l<<-l+1\n",
    "            # determine birth rate, number of driver/passenger mutations, and affected genes of new clone\n",
    "            tmp1 <- count_hits(node, exann, lapply(simout,'[',i)[[1]], lapply(simout,'[',i)[[2]], sd, sp)\n",
    "            # determine the loci/L1s of insertions in the new clone\n",
    "            tmp2 <- mapply(append, lapply(simout,'[',i), node$tes, SIMPLIFY = FALSE)\n",
    "            node$AddChild(l, ncells=1, B=tmp1[[1]], nd=node$nd+tmp1[[3]], np=node$np+tmp1[[2]], genes=rbind(node$genes,tmp1[[4]]), tes=tmp2)\n",
    "            gn <<- rbind(gn,tmp1[[4]]) # append hit genes to list\n",
    "        }\n",
    "    }   \n",
    "    node$ncells <- append(node$ncells,nc)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--- Set simulation parameters\n",
    "ENifrc<- .1       # Fraction of endonuclease-independent (random) insertions\n",
    "N0 <- 1           # Initial number of cells in root clone\n",
    "B0 <- 1           # Initial birth rate\n",
    "mu <- 0.2           # Probability of transposition / timestep of a single cell\n",
    "\n",
    "NT <- 5           # Number of time steps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clone tree creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.998002996004994"
      ],
      "text/latex": [
       "0.998002996004994"
      ],
      "text/markdown": [
       "0.998002996004994"
      ],
      "text/plain": [
       "[1] 0.998003"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>elapsed:</strong> 0.348999999999933"
      ],
      "text/latex": [
       "\\textbf{elapsed:} 0.348999999999933"
      ],
      "text/markdown": [
       "**elapsed:** 0.348999999999933"
      ],
      "text/plain": [
       "elapsed \n",
       "  0.349 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          levelName     ncells        B\n",
      "1 1                       1, 0 0.998003\n",
      "2  ¦--2             1, 2, 3, 5 0.998003\n",
      "3  ¦   ¦--5               1, 2 0.998003\n",
      "4  ¦   °--7                  1 0.998003\n",
      "5  °--3             1, 1, 2, 4 0.998003\n",
      "6      °--4            1, 1, 1 0.998003\n",
      "7          ¦--6           1, 1 0.998003\n",
      "8          ¦   °--9          1 0.998003\n",
      "9          °--8              1 0.998003\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>1</li>\n",
       "\t<li>2</li>\n",
       "\t<li>4</li>\n",
       "\t<li>8</li>\n",
       "\t<li>16</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 4\n",
       "\\item 8\n",
       "\\item 16\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "3. 4\n",
       "4. 8\n",
       "5. 16\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]  1  2  4  8 16"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>sym</th><th scope=col>tsg</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>DDX11L1</td><td>0      </td></tr>\n",
       "\t<tr><td>MTOR-AS1, MTOR    </td><td>0                 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " sym & tsg\\\\\n",
       "\\hline\n",
       "\t DDX11L1 & 0      \\\\\n",
       "\t MTOR-AS1, MTOR     & 0                 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "sym | tsg | \n",
       "|---|---|\n",
       "| DDX11L1 | 0       | \n",
       "| MTOR-AS1, MTOR     | 0                  | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  sym                tsg\n",
       "1 DDX11L1            0  \n",
       "2 MTOR-AS1, MTOR     0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "l<-1 # Clone counter\n",
    "\n",
    "N <- N0 # total population size\n",
    "gn <- c() # list of genes hit\n",
    "\n",
    "CellPop <- Node$new(1) # Initialize data.tree as single node\n",
    "\n",
    "CellPop$ncells <- c(N0) # Set initial number of cells of clone\n",
    "CellPop$B <- B0 # Set initial birth rate of clone\n",
    "CellPop$np <- 0 # Set initial number of drivier mutations\n",
    "CellPop$nd <- 0 # Set initial number of passenger mutations\n",
    "CellPop$genes <- c()\n",
    "# Set the L1 insertions existing in root node\n",
    "CellPop$tes <- list(c(),c(),c(),c())\n",
    "CellPop$tes <- list(c(\"chr1\",\"chr1\"),c(11874,11149270),c(\"+\",\"+\"),c(0,0))\n",
    "\n",
    "# update attributes based on initialized insertions\n",
    "tmp <- count_hits(CellPop, exann, CellPop$tes[[1]], CellPop$tes[[2]], 0.1, 0.001)\n",
    "CellPop$B <- tmp[[1]]\n",
    "CellPop$np <- tmp[[2]]\n",
    "CellPop$nd <- tmp[[3]]\n",
    "CellPop$genes <- tmp[[4]]\n",
    "gn <- tmp[[4]] # list of genes hit\n",
    "\n",
    "# print birth rate of root clone\n",
    "CellPop$B\n",
    "\n",
    "ptm <- proc.time()\n",
    "for (i in 2:NT) {\n",
    "\n",
    "    t <- Traverse(CellPop,traversal='pre-order',filterFun=function(x) tail(x$ncells,n=1) > 0)    \n",
    "    lapply(t,maybeTranspose,0.1,0.001)\n",
    "\n",
    "    N <- append(N,sum(vapply(CellPop$Get('ncells'),tail,n=1L,FUN.VALUE = numeric(1))))\n",
    "\n",
    "    \n",
    "}\n",
    "(proc.time() - ptm)[3]\n",
    "\n",
    "print(CellPop,'ncells','B')\n",
    "# sum(vapply(CellPop$Get('ncells'),tail,n=1L,FUN.VALUE = numeric(1)))\n",
    "# CellPop$Get('ncells')\n",
    "N\n",
    "gn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
