{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: XVector\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "Loading required package: BSgenome\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: rtracklayer\n",
      "Warning message:\n",
      "“replacing previous import ‘BiocGenerics::dims’ by ‘Biobase::dims’ when loading ‘SummarizedExperiment’”Warning message:\n",
      "“replacing previous import ‘Biobase::dims’ by ‘DelayedArray::dims’ when loading ‘SummarizedExperiment’”"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading reference genome (GRCh38)...\n"
     ]
    }
   ],
   "source": [
    "# -----------------------------------------------------------------------------\n",
    "# map_genome.r\n",
    "\n",
    "# Maps potential insertion sites based on the 'Snap-Velcro' model described \n",
    "# in Clement Monot, et al. (2013) \"The Specificity and Flexibility of L1 Reverse \n",
    "# Transcription Priming at Imperfect T-Tracts.\" PLOS Genetics, 9:5.\n",
    "#\n",
    "# This script creates a separate R data file for each chromosome containing the \n",
    "# locations of each type of EN site for both sense and anti-sense strands.\n",
    "# Locations and scores of the putative EN sites are initially stored in memory. \n",
    "#\n",
    "# Usage: ./mapgenome.r <start chromosome # (e.g. 1)> <end chromosome # (e.g. 24)>\n",
    "# Output: <chromosome name>map.rda to data directory file specified on line 129.\n",
    "#\n",
    "# Dependencies: R(>= 2.8.0, Packages - Biostrings, BSgenome (for default hg38), \n",
    "# GenomicRanges)\n",
    "# -----------------------------------------------------------------------------\n",
    "\n",
    "#--- Load libraries\n",
    "library(Biostrings)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "map<-c()\n",
    "\n",
    "start<-1\n",
    "end<-24\n",
    "options(warn=1)\n",
    "\n",
    "#--- hg38 reference genome\n",
    "cat(\"\\nReading reference genome (GRCh38)...\\n\")\n",
    "genome<-Hsapiens\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chromosome:  chr1 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 1.348154 mins\n",
      "\n",
      "Chromosome:  chr2 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 1.370992 mins\n",
      "\n",
      "Chromosome:  chr3 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 1.132926 mins\n",
      "\n",
      "Chromosome:  chr4 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 1.167382 mins\n",
      "\n",
      "Chromosome:  chr5 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 1.012627 mins\n",
      "\n",
      "Chromosome:  chr6 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 47.34319 secs\n",
      "\n",
      "Chromosome:  chr7 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 38.98664 secs\n",
      "\n",
      "Chromosome:  chr8 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 35.25804 secs\n",
      "\n",
      "Chromosome:  chr9 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 28.81626 secs\n",
      "\n",
      "Chromosome:  chr10 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 32.14648 secs\n",
      "\n",
      "Chromosome:  chr11 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 32.36239 secs\n",
      "\n",
      "Chromosome:  chr12 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 33.77079 secs\n",
      "\n",
      "Chromosome:  chr13 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 29.48657 secs\n",
      "\n",
      "Chromosome:  chr14 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 19.99595 secs\n",
      "\n",
      "Chromosome:  chr15 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 15.99787 secs\n",
      "\n",
      "Chromosome:  chr16 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 15.9317 secs\n",
      "\n",
      "Chromosome:  chr17 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 14.12002 secs\n",
      "\n",
      "Chromosome:  chr18 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 21.24744 secs\n",
      "\n",
      "Chromosome:  chr19 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 7.493015 secs\n",
      "\n",
      "Chromosome:  chr20 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 12.06074 secs\n",
      "\n",
      "Chromosome:  chr21 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 8.876798 secs\n",
      "\n",
      "Chromosome:  chr22 \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 5.827769 secs\n",
      "\n",
      "Chromosome:  chrX \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 42.55246 secs\n",
      "\n",
      "Chromosome:  chrY \n",
      "Creating EN site annotation: sense...\n",
      "Creating EN site annotation: anti-sense...\n",
      "Saving map file...\n",
      "Time difference of 5.981868 secs\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# chrnm <- \"chr1\"\n",
    "\n",
    "#--- For each chromosome in hg38 range of <start> to <end>...\n",
    "for (chrnm in names(genome)[start:end]) {\n",
    "\tstart.time<-Sys.time()\n",
    "\n",
    "\t### Sense strand analysis\n",
    "\t###------------------------------------------------------------------------###\n",
    "\tcat(\"\\nChromosome: \",chrnm,\"\\n\")\t\n",
    "\tcat(\"Creating EN site annotation: sense...\\n\")\n",
    "    \n",
    "    flush.console()\n",
    "\n",
    "\t#--- 5' to 3' target sequence 'TTTT' (derived from Monot, et. al)\n",
    "\ttar <- DNAString(\"TTTTA\")\n",
    "\t#--- Finds target starting location in chromosome, one mismatch is allowed\n",
    "\tmtchView <- matchPattern(tar,genome[[chrnm]],max.mismatch=1)\n",
    "    purines <- DNAStringSet(genome[[chrnm]],end(mtchView),end(mtchView))\n",
    "    mtchView <- mtchView[purines=='A'] # Doesn't allow mismatches for the purine\n",
    "\t#--- Calculates position weighted T-density scores\n",
    "\tprimrngs <- IRanges(start=start(mtchView)-6,width=6) # start and end points of 6 bp upstream of targets\n",
    "\ttargets <- DNAStringSet(genome[[chrnm]],start(mtchView),end(mtchView)) # collect targets from genome\n",
    "\tprmrs <- DNAStringSet(genome[[chrnm]],start(primrngs),end(primrngs)) # collect 'primers' - 6 bp upstream of each target\n",
    "\t#--- Create a list of lists containing locations of Ts in each primer\n",
    "\ttmp <- vmatchPattern(\"T\",prmrs)\n",
    "\ttmp <- startIndex(tmp)\n",
    "\t#--- 0.8456 is the maximum velcro score, corresponding to 'TTTTTT' (6).\n",
    "\tprimrnks <- lapply(tmp,function(x) sum(1/(11-x))/0.84563492) \n",
    "\trm(tmp) # cleans up\n",
    "\n",
    "    #--- Stores indices of sites for each target category (from Monot, et. al).\n",
    "    #--- Example: 'ict_s' is site for 'closed' and 'tight' insertion.\n",
    "    ict_s<-which(targets==tar & primrnks >= 0.5)\n",
    "    icl_s<-which(targets==tar & primrnks < 0.5)\n",
    "    iot_s<-which(targets!=tar & primrnks >= 0.5)\n",
    "    iol_s<-which(targets!=tar & primrnks < 0.5)\n",
    "    insites_s<-end(mtchView)\n",
    "\n",
    "\t### Anti-sense strand analysis\n",
    "\t###------------------------------------------------------------------------###\n",
    "\tcat(\"Creating EN site annotation: anti-sense...\\n\")\n",
    "\n",
    "\t#--- 3' to 5' anti-sense target sequence 'AAAA' (derived from Monot, et. al)\n",
    "\ttar <- DNAString(\"TAAAA\")\n",
    "\tmtchView <- matchPattern(tar,genome[[chrnm]],max.mismatch=1)\n",
    "    purines <- DNAStringSet(genome[[chrnm]],start(mtchView), start(mtchView))\n",
    "    mtchView <- mtchView[purines=='T'] # Doesn't allow mismatches for the purine\n",
    "    primrngs <- IRanges(start=end(mtchView)+1,width=6)\n",
    "\ttargets <- DNAStringSet(genome[[chrnm]],start(mtchView),end(mtchView))\n",
    "    prmrs <- DNAStringSet(genome[[chrnm]],start(primrngs),end(primrngs))\n",
    "    tmp <- vmatchPattern(\"A\",prmrs)\n",
    "    tmp <- startIndex(tmp)\n",
    "    primrnks <- lapply(tmp,function(x) sum(1/(x+4))/0.84563492)\n",
    "    rm(tmp)\n",
    "    rm(purines)\n",
    "\n",
    "\tict_ns<-which(targets==tar & primrnks >= 0.5)\n",
    "\ticl_ns<-which(targets==tar & primrnks < 0.5)\n",
    "\tiot_ns<-which(targets!=tar & primrnks >= 0.5)\n",
    "\tiol_ns<-which(targets!=tar & primrnks < 0.5)\n",
    "\tinsites_ns<-start(mtchView)\n",
    "\n",
    "\t### Concatenate data from both strands\n",
    "\t###-----------------------------------------------------------------------##\n",
    "\n",
    "\t#--- Make the arrays for each set of indices equal length, padding the shorter with NAs\n",
    "\tn <- max(length(ict_s), length(ict_ns))\n",
    "\tlength(ict_s) <- n                      \n",
    "\tlength(ict_ns) <- n\n",
    "    n <- max(length(icl_s), length(icl_ns))\n",
    "\tlength(icl_s) <- n\n",
    "    length(icl_ns) <- n\n",
    "    n <- max(length(iot_s), length(iot_ns))\n",
    "    length(iot_s) <- n\n",
    "    length(iot_ns) <- n\n",
    "    n <- max(length(iol_s), length(iol_ns))\n",
    "    length(iol_s) <- n\n",
    "    length(iol_ns) <- n\t\n",
    "    n <- max(length(insites_s), length(insites_ns))\n",
    "    length(insites_s) <- n\n",
    "    length(insites_ns) <- n\n",
    "\t#--- Bind the arrays for both strands as two columns\n",
    "\tict<-cbind(ict_s,ict_ns)\n",
    "\ticl<-cbind(icl_s,icl_ns)\n",
    "\tiot<-cbind(iot_s,iot_ns)\n",
    "\tiol<-cbind(iol_s,iol_ns)\n",
    "\tinsites<-cbind(insites_s,insites_ns)\n",
    "\t#--- Store data in single S4 object\n",
    "\tmap$insites <-insites\n",
    "\tmap$ict<-ict\n",
    "\tmap$icl<-icl\n",
    "\tmap$iot<-iot\n",
    "\tmap$iol<-iol\n",
    "\n",
    "\t#--- Save map under name of current chromosome\n",
    "\tassign(paste0(chrnm,\"Map\"),map)\n",
    "\n",
    "\tcat(\"Saving map file...\\n\")\n",
    "\twd <- getwd()\n",
    "\tsave(list=(paste0(chrnm,\"Map\")),file=paste(wd, \"/data/l1_site_maps/\",chrnm,\".rda\",sep=\"\"))\n",
    "\trm(list=c(\"insites\",\"ict\",\"icl\",\"iot\",\"iol\"))\n",
    "\tend.time <- Sys.time()\n",
    "    print(end.time-start.time)\n",
    "    \n",
    "    flush.console()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
