{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load libraries and data files\n",
    "###############################################################\n",
    "library(data.table)\n",
    "load('../data/exonicvsnon_counts.rda') # Probabilities of each mutation type\n",
    "load('../data/gene_pdt.rda') # Load tables giving probability of insertion for each gene in a diploid genome for both genders\n",
    "xy_genes <- gene_pdt_m$geneSym[gene_pdt_m$chrom %in% c('X','Y')] # Getting list of genes on chroms. X and Y by symbol\n",
    "\n",
    "# Define functions\n",
    "###############################################################\n",
    "##### birthrate() #####\n",
    "birthrate <- function(nd_het, np_het, nd_hom, np_hom, sld, slp, spd, spp) { # cell cycles per time step per cell \n",
    "    return(((((1+sld)^nd_het)*((1+spd)^nd_hom))/(((1+slp)^np_het)*((1+spp)^np_hom))))\n",
    "}\n",
    "\n",
    "##### delta_ncells() #####\n",
    "delta_ncells <- function(B, D, ncells, tau, genTime) { # change in number of cells\n",
    "    return (max(ncells + rpois(1,ncells*B*tau*genTime)-rpois(1,ncells*D),0))\n",
    "}\n",
    "\n",
    "##### get_mu_i() #####\n",
    "get_mu_i <- function(B, mu, tau, genTime) {return(mu*B*tau*genTime)} # insertions per time step per cell\n",
    "\n",
    "##### get_nins() #####\n",
    "get_nins <- function(ncells, mu_i) {return (rpois(1,ncells*mu_i))} # number of insertions in clone\n",
    "\n",
    "##### get_drivers() #####\n",
    "get_drivers <- function(cancer_type){\n",
    "    \n",
    "    geneList <- read.csv('../data/Census_allThu Sep 13 17_30_17 2018.csv') # Cancer Gene Census Data Table\n",
    "    if (cancer_type=='lung') {\n",
    "        toMatch <- c('lung')\n",
    "    } else if (cancer_type=='colon') {\n",
    "        toMatch <- c('colon','colorectal')\n",
    "    } else if (cancer_type=='brain') {\n",
    "        toMatch <- c('astrocytoma','meningioma','head','oligodendroglio')\n",
    "    }\n",
    "    # Get genes affecting chosen cancer type\n",
    "    geneList<-geneList[grep(paste(toMatch,collapse='|'),geneList$Tumour.Types.Somatic),]\n",
    "    # Filter for only Tier 1 genes\n",
    "#     geneList<-geneList[geneList$Tier==1,]\n",
    "    # Filter for TSGs\n",
    "    geneList<-geneList[grep('TSG',geneList$Role.in.Cancer),]\n",
    "    geneList <- as.vector(geneList$Gene.Symbol)\n",
    "    \n",
    "    return(geneList)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### update_genes_f() #####\n",
    "# This function will update the gene lists for a given clone, for a female genome\n",
    "# types 1 - no effect\n",
    "# types 2 - heterozygous effect\n",
    "# types 3 - homozygous effect\n",
    "update_genes_f <- function(new_genes,genes_het,genes_hom) {\n",
    "    \n",
    "    types <- rep(0,length(new_genes))\n",
    "    jj <- 1\n",
    "    for (ii in new_genes) {\n",
    "        if (ii %in% genes_hom) { # If the new gene is in list of homozygous genes, ignore it\n",
    "            types[jj]<-1\n",
    "        } else {\n",
    "            if (ii %in% genes_het) { # If the new gene is in the list of heterozygous genes\n",
    "                if (runif(1,c(0,1))>=0.5) { # With 50% probability, assume it's the other copy, add the gene to the homoz. list, and discard it from heteroz. list\n",
    "                    genes_hom <- append(genes_hom,ii)\n",
    "                    genes_het <- genes_het[genes_het!=ii]\n",
    "                    types[jj]<-3\n",
    "                } else { # Else, assume it hit already disrupted copy, and discard from list\n",
    "                    types[jj]<-1\n",
    "                }\n",
    "            } else {\n",
    "                genes_het <- append(genes_het,ii)\n",
    "                types[jj]<-2\n",
    "            }\n",
    "        }\n",
    "        jj <- jj+1\n",
    "    }\n",
    "    return(list(genes_het,genes_hom,list(''),types))\n",
    "    \n",
    "}\n",
    "\n",
    "##### update_genes_m() #####\n",
    "# This function will update the gene lists for a given clone, for a male genome\n",
    "# type 1 - no effect\n",
    "# type 2 - heterozygous effect\n",
    "# type 3 - homozygous effect\n",
    "# type 4 - X/Y gene effect\n",
    "update_genes_m <- function(new_genes,genes_het,genes_hom) {\n",
    "    \n",
    "    types <- rep(0,length(new_genes))\n",
    "    jj <- 1\n",
    "    for (ii in new_genes) {\n",
    "        if (ii %in% genes_hom) { # If the new gene is in the list of homoz. genes, ignore it\n",
    "            types[jj] <-1\n",
    "        } else if (ii %in% xy_genes) { # If not in homoz. gene list, but in X/Y gene list\n",
    "            genes_hom <- append(genes_hom,ii)\n",
    "            types[jj] <- 4 # FIX ME\n",
    "        } else {\n",
    "            if (ii %in% genes_het) { # If the new gene is in the list of heteroz. genes\n",
    "                \n",
    "                if (runif(1,c(0,1))>=0.5) { # With 50% probability, assume it's the other allele, add the gene to the homoz. list, and discard it from heteroz. list\n",
    "                    genes_hom <- append(genes_hom,ii)\n",
    "                    genes_het <- genes_het[genes_het!=ii]\n",
    "                    types[jj]<-3\n",
    "                } else { # Else, assume it hit already disrupted copy, and discard from list\n",
    "                    types[jj]<-1\n",
    "                }\n",
    "                \n",
    "            } else {\n",
    "                genes_het <- append(genes_het,ii)\n",
    "                types[jj]<-2\n",
    "            }\n",
    "        }\n",
    "        jj <- jj+1\n",
    "    }\n",
    "    return(list(genes_het,genes_hom,list(''),types))   \n",
    "    \n",
    "}    \n",
    "\n",
    "##### update_mcount_f() #####\n",
    "update_mcount_f <- function(nd_het,np_het,nd_hom,np_hom,typemut,typegene) {\n",
    "    \n",
    "    for (ii in 1:length(typemut)) {\n",
    "        if (typemut[ii]==2 & typegene[ii]==0) { # If heterozygous passenger mutation (most likely)\n",
    "            np_het <- np_het+1\n",
    "        } else if (typemut[ii]==1) { # If no-effect (redundant) mutation\n",
    "        } else if (typemut[ii]==2 & typegene[ii]==1) { # If heterozygous driver mutation \n",
    "            nd_het <- nd_het+1\n",
    "        } else if (typemut[ii]==3 & typegene[ii]==0) { # If homozygous passenger\n",
    "            np_het <- np_het-1\n",
    "            np_hom <- np_hom+1\n",
    "        } else if (typemut[ii]==3 & typegene[ii]==1) { # If homozygous driver\n",
    "            nd_het <- nd_het-1\n",
    "            nd_hom <- nd_hom+1\n",
    "        }\n",
    "    }\n",
    "    return(list(nd_het,np_het,nd_hom,np_hom))\n",
    "}\n",
    "\n",
    "##### update_mcount_m() #####\n",
    "update_mcount_m <- function(nd_het,np_het,nd_hom,np_hom,typemut,typegene) {\n",
    "    \n",
    "    for (ii in 1:length(typemut)) {\n",
    "        if (typemut[ii]==2 & typegene[ii]==0) { # If heterozygous passenger mutation (most likely)\n",
    "            np_het <- np_het+1\n",
    "        } else if (typemut[ii]==1) { # If no-effect (redundant) mutation\n",
    "        } else if (typemut[ii]==2 & typegene[ii]==1) { # If heterozygous driver mutation \n",
    "            nd_het <- nd_het+1\n",
    "        } else if (typemut[ii]==3 & typegene[ii]==0) { # If homozygous passenger\n",
    "            np_het <- np_het-1\n",
    "            np_hom <- np_hom+1\n",
    "        } else if (typemut[ii]==3 & typegene[ii]==1) { # If homozygous driver\n",
    "            nd_het <- nd_het-1\n",
    "            nd_hom <- nd_hom+1\n",
    "        } else if (typemut[ii]==4 & typegene[ii]==0) { # If X/Y passenger\n",
    "            np_hom <- np_hom+1\n",
    "        } else if (typemut[ii]==4 & typegene[ii]==1) { # If X/Y driver\n",
    "            nd_hom <- nd_hom+1\n",
    "        }\n",
    "    }\n",
    "    return(list(nd_het,np_het,nd_hom,np_hom))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### run_sim() #####\n",
    "run_sim <- function(N0, mu, tau, NT, sld, slp, spd, spp, gender, driverGene, geneList, nclones, logpath) {\n",
    "\n",
    "    if (gender=='male') {\n",
    "        gene_pdt <- gene_pdt_m\n",
    "        update_genes <- update_genes_m\n",
    "        update_mcount <- update_mcount_m\n",
    "    } else if (gender=='female') {\n",
    "        gene_pdt <- gene_pdt_f\n",
    "        update_genes <- update_genes_f\n",
    "        update_mcount <- update_mcount_f\n",
    "    } else {stop('Argument gender must be \\'male\\' or \\'female\\'.')}\n",
    "    \n",
    "    gene_pdt$type[gene_pdt$geneSym %in% geneList] <- 1\n",
    "    write(paste0(toString(length(which(gene_pdt$type==1))),' out of ',length(geneList),' driver genes found in gene_pdt.'),file=logpath,append=TRUE)\n",
    "    \n",
    "    # Allocate population\n",
    "    Pop <- data.table(ncells=rep(0,nclones),\n",
    "                      B=rep(0,nclones),\n",
    "                      mu_i=rep(0,nclones),\n",
    "                      nd_het=rep(0,nclones),\n",
    "                      np_het=rep(0,nclones),\n",
    "                      nd_hom=rep(0,nclones),\n",
    "                      np_hom=rep(0,nclones),\n",
    "                      genes_het=rep(list(''),nclones),\n",
    "                      genes_hom=rep(list(''),nclones),\n",
    "                      genes_new=rep(list(''),nclones),\n",
    "                      new_types=rep(list(),nclones))\n",
    "    \n",
    "    # Initialize population with a heterozygous driver mutation\n",
    "#     Pop[1:2,c('ncells','nd_het','np_het','nd_hom','np_hom','genes_het','genes_hom','genes_new','new_types'):=list(c(N0-1,1),\n",
    "#                                                                                  c(0,1),\n",
    "#                                                                                  c(0,0),\n",
    "#                                                                                  c(0,0),\n",
    "#                                                                                  c(0,0),\n",
    "#                                                                                  list(c(''),c(driverGene)),\n",
    "#                                                                                  list(c(''),c('')),\n",
    "#                                                                                  list(c(''),c('')),\n",
    "#                                                                                  c(0,1))]\n",
    "    \n",
    "    # Initialize population with no mutations\n",
    "    Pop[1,c('ncells','nd_het','np_het','nd_hom','np_hom','genes_het','genes_hom','genes_new','new_types'):=list(c(N0),\n",
    "                                                                                 c(0),\n",
    "                                                                                 c(0),\n",
    "                                                                                 c(0),\n",
    "                                                                                 c(0),\n",
    "                                                                                 list(c('')),\n",
    "                                                                                 list(c('')),\n",
    "                                                                                 list(c('')),\n",
    "                                                                                 list(c()))]\n",
    "    \n",
    "    # Assign birth and insertion rates\n",
    "    Pop[1:2, B := mapply(birthrate, nd_het, np_het, nd_hom, np_hom, sld, slp, spd, spp)]\n",
    "    Pop[1:2, mu_i := mapply(get_mu_i, B, mu, tau, 1)]\n",
    "    \n",
    "    N <- rep(0,NT) # Allocate array for population size time series\n",
    "    genTime <- rep(0,NT) # Allocate array for generation time factor\n",
    "    genes <- character(nrow(gene_pdt))\n",
    "    write('Initialized...',file=logpath,append=TRUE)\n",
    "\n",
    "    ptm <- proc.time()\n",
    "    for (ii in 1:NT) { # Loop over time steps\n",
    "        if(ii %in% c(round(NT/4),round(NT/4*2),round(NT/4*3),NT)) { # Print progress at 25% completed intervals\n",
    "            write(paste0(toString(ii/NT*100),'% done | ',format((proc.time()-ptm)[1],nsmall=3),' (s)'),file=logpath,append=TRUE)            \n",
    "        }\n",
    "        \n",
    "        clog <- Pop$ncells>0 # Get logical array for indices of active (# cells >0) clones\n",
    "        \n",
    "        N[ii] <- sum(Pop$ncells) # Get current number of cells\n",
    "        if (N[ii]>=3*N0 || N[ii]<1) {break} # Simulation stops if population has grown by 3X or died\n",
    "        genTime[ii] <- 1/mean(Pop$B[clog]) # Get generation length\n",
    "        D <- N[ii]*tau*genTime[ii]/N0 # Compute death rate (cell deaths per time step per cell)\n",
    "\n",
    "        nins <- sum(unlist(mapply(get_nins,Pop$ncells[clog],Pop$mu_i[clog],SIMPLIFY=FALSE))) # Get number of exonic insertions\n",
    "        if (nins > 0) {\n",
    "            \n",
    "            rownew <- which(Pop$ncells==0)[1] # Find first row of the data table with ncells==0\n",
    "            \n",
    "            # Sample cells for mutations, with replacement, with probability determined by mu\n",
    "            cellsWIns <- table(sample(1:sum(Pop$ncells[clog]), nins, replace=TRUE, prob=rep(Pop$mu_i[clog], Pop$ncells[clog])))\n",
    "            clonesWIns <- rep(1:length(which(clog)), Pop$ncells[clog])[as.integer(names(cellsWIns))] # Get clone ID for each cell with mutation(s)\n",
    "            ctab <- table(clonesWIns)\n",
    "            cids <- as.integer(names(ctab)) # Get row ids of sampled clones\n",
    "            set(Pop,cids,1L,Pop[cids,1L] - as.integer(ctab)) # Remove cells from sampled clones\n",
    "            \n",
    "            # Sample genes for mutations, with replacement, with probability determined by target site distribution\n",
    "            gene_ids <- sample(1:nrow(gene_pdt),nins,replace=TRUE,prob=gene_pdt$p)\n",
    "            gene_list <- gene_pdt$geneSym[gene_ids]\n",
    "            genes <- append(genes,gene_list)\n",
    "            gene_types <- gene_pdt$type[gene_ids]\n",
    "            \n",
    "            # Get list of genes for each cell\n",
    "            tmp1 <- rep(list(),length(cellsWIns))\n",
    "            tmp2 <- rep(list(),length(cellsWIns))\n",
    "            for (jj in 1:length(cellsWIns)) {\n",
    "                tmp1[[jj]] <- head(gene_list,as.integer(cellsWIns)[jj])\n",
    "                tmp2[[jj]] <- head(gene_types,as.integer(cellsWIns)[jj])\n",
    "                gene_list <- tail(gene_list,length(gene_ids)-as.integer(cellsWIns)[jj])\n",
    "                gene_types <- tail(gene_types,length(gene_types)-as.integer(cellsWIns)[jj])\n",
    "            }\n",
    "            gene_list <- tmp1\n",
    "            gene_types <- tmp2\n",
    "            \n",
    "            new_inds <- rownew:(rownew+length(clonesWIns)-1)\n",
    "            Pop[new_inds, c(\"ncells\",\"nd_het\",\"np_het\",\"nd_hom\",\"np_hom\",\"genes_het\",\"genes_hom\",\"genes_new\",\"new_types\"):=list(1, \n",
    "                                                                                        Pop$nd_het[clonesWIns], \n",
    "                                                                                        Pop$np_het[clonesWIns], \n",
    "                                                                                        Pop$nd_hom[clonesWIns], \n",
    "                                                                                        Pop$np_hom[clonesWIns], \n",
    "                                                                                        Pop$genes_het[clonesWIns],\n",
    "                                                                                        Pop$genes_hom[clonesWIns],\n",
    "                                                                                        gene_list,\n",
    "                                                                                        gene_types)]\n",
    "            \n",
    "            # Update gene lists\n",
    "            tmp1 <- t(mapply(update_genes,Pop$genes_new[new_inds],Pop$genes_het[new_inds],Pop$genes_hom[new_inds],SIMPLIFY=TRUE))\n",
    "            Pop$genes_het[new_inds] <- tmp1[,1]\n",
    "            Pop$genes_hom[new_inds] <- tmp1[,2]\n",
    "            Pop$genes_new[new_inds] <- tmp1[,3]\n",
    "            \n",
    "            # Update insertion counts\n",
    "            tmp2<-t(mapply(update_mcount,Pop$nd_het[new_inds],Pop$np_het[new_inds],Pop$nd_hom[new_inds],Pop$np_hom[new_inds],tmp1[,4],Pop$new_types[new_inds]))\n",
    "            Pop[new_inds,c(\"nd_het\",\"np_het\",\"nd_hom\",\"np_hom\"):=list(unlist(tmp2[,1]),unlist(tmp2[,2]),unlist(tmp2[,3]),unlist(tmp2[,4]))]\n",
    "            \n",
    "            # Update birth and insertion rates\n",
    "            Pop[new_inds, B := mapply(birthrate, nd_het, np_het, nd_hom, np_hom, sld, slp, spd, spp)]\n",
    "            Pop[new_inds, mu_i := mapply(get_mu_i, B, mu, tau, genTime[ii])]\n",
    "        }\n",
    "        \n",
    "        Pop[Pop$ncells>0, ncells:=mapply(delta_ncells, B, D, ncells, tau, genTime[ii])] # Update number of cells for all clones\n",
    "        Pop <- Pop[order(Pop$ncells,decreasing=TRUE),] # Order data.table by ncells\n",
    "    }\n",
    "    print(proc.time() - ptm)\n",
    "    genes <- genes[!is.na(genes)]\n",
    "    Pop <- Pop[,1:9]\n",
    "    return(list(Pop,N,genes,genTime))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################################3\n",
    "# Gender of organism\n",
    "gender <- 'male'\n",
    "\n",
    "# Symbol of driver gene disrupted at t=0\n",
    "driverGene <- ''\n",
    "\n",
    "# List of driver genes\n",
    "geneList <- get_drivers('lung')\n",
    "\n",
    "# Initial number of cells\n",
    "N0 <- 1000\n",
    "\n",
    "# Average # L1 insertions / cell cycle\n",
    "mu <- 0.05\n",
    "if (gender=='male') {\n",
    "    mu <- mu*(1-pd_exvsnon_m[2])# Scale by 1-probability of null (intron) insertion, to convert to probability of exonic insertion\n",
    "} else if (gender=='female') {\n",
    "    mu <- mu*(1-pd_exvsnon_f[2])\n",
    "}\n",
    "\n",
    "# Selective strengths of each mutation type\n",
    "spd <- .1 # Homozygous driver\n",
    "spp <- .005 # Homozygous passenger\n",
    "sld <- spd*0.1 # Heterozygous driver\n",
    "slp <- spp*0.1 # Hetero. passenger\n",
    "\n",
    "# Number of time steps to simulate\n",
    "NT <- 2e4/4\n",
    "\n",
    "###################################################3\n",
    "# Time resolution: number of timesteps per generation\n",
    "tau <- 1\n",
    "tau <- 1/tau\n",
    "# Buffer size of population data object; represents the max possible number of clones in the population\n",
    "maxNClones <- N0*4\n",
    "# Log path\n",
    "logpath <- './test_v2.4.log'\n",
    "system(paste0('rm ',logpath))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out <- run_sim(N0, mu, tau, NT, sld, slp, spd, spp, gender, driverGene, geneList, maxNClones, logpath)\n",
    "Pop <- out[[1]]\n",
    "N <- out[[2]]\n",
    "genTime <- out[[4]]\n",
    "\n",
    "# head(Pop)\n",
    "N[N==0] <- NA\n",
    "genTime[genTime==0] <- NA\n",
    "plot(1:NT*tau,N,type='l',xlab='Generation',ylab='Population Size')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Batch run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nrun <- 0\n",
    "\n",
    "N0v <- c(1e3)\n",
    "muv <- c(.05)\n",
    "if (gender=='male') {\n",
    "    muv_adj <- muv*(1-pd_exvsnon_m[2])\n",
    "} else if (gender=='female') {\n",
    "    muv_adj <- muv*(1-pd_exvsnon_f[2])\n",
    "}\n",
    "sdv <- c(.1)\n",
    "spv <- c(.005)\n",
    "hef <- 0.1 # Fraction of homozygous effect assigned to heterozygous mutations\n",
    "\n",
    "outPath <- '../../lung_1017/'\n",
    "logPath <- paste0(outPath,'run_1.log')\n",
    "\n",
    "tryCatch({\n",
    "    for (ii in 1:length(N0v)){\n",
    "        for (jj in 1:length(muv)) {\n",
    "            for (ll in 1:length(sdv)) {\n",
    "                for (mm in 1:length(spv)) {\n",
    "                    for (nn in 1:500) {\n",
    "                        nrun <- nrun+1\n",
    "                        line <- paste0('Run: ',toString(nrun),'\\tN0: ',toString(N0v[ii]),'\\tmu: ',toString(muv[jj]),'\\tsd: ',toString(sdv[ll]),'\\tsp: ',toString(spv[mm]))\n",
    "                        write(line,file=logPath,append=TRUE)\n",
    "                        out <- run_sim(N0v[ii], muv_adj[jj], tau, NT, sdv[ll]*hef, spv[mm]*hef, sdv[ll], spv[mm], gender, driverGene, geneList, N0v[ii]*4, logPath)\n",
    "                        Pop <- out[[1]]\n",
    "                        N <- out[[2]]\n",
    "                        save(Pop,N,file=paste0(outPath,'run_n0',N0v[ii],'_',nrun,\".rda\"))\n",
    "                        rm(Pop,N)\n",
    "                    }\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    }, warning = function(warn.msg){\n",
    "        write(toString(warn.msg), logPath, append=TRUE)\n",
    "    }, error = function(err.msg){\n",
    "        write(toString(err.msg), logPath, append=TRUE)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
